{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "nbsphinx": "hidden"
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       "\n",
       "    #jupman-nav {\n",
       "        position: fixed;\n",
       "        top: 10px; \n",
       "        left: 5px; \n",
       "        background: white;\n",
       "        width: 200px; \n",
       "        padding: 10px; \n",
       "        z-index: 10;\n",
       "    }\n",
       "\n",
       "    /* style=\"background: rgba(255, 255, 255, 0);\" class=\"unstyled\" */\n",
       "  \n",
       "    #jupman-toc {\n",
       "        position: fixed;\n",
       "        top: 100px; \n",
       "        left: -55px; \n",
       "        background: rgba(255, 255, 255, 0);\n",
       "        \n",
       "        height: 80%;\n",
       "        width: 200px; \n",
       "        padding: 20px; \n",
       "        z-index: 10;\n",
       "        overflow-y: scroll;        \n",
       "            /* font-size: 50%; */\n",
       "    }\n",
       "\n",
       "    \n",
       "    \n",
       "    code {\n",
       "        background-color:rgb(245, 249, 249) !important;\n",
       "    }\n",
       "    \n",
       "    \n",
       "    .jupman-nav-item {\n",
       "        font-weight:bold;\n",
       "        font-size:120%  \n",
       "    }\n",
       "    \n",
       "    .jupman-chapter-link {\n",
       "        font-size:150%;\n",
       "        margin-right:15px;        \n",
       "    }\n",
       "    \n",
       "    .jupman-chapter-link a {\n",
       "        text-decoration: none;\n",
       "    }    \n",
       "    \n",
       "    .jupman-title {\n",
       "        font-size:250%; \n",
       "        font-weight:bold        \n",
       "    }    \n",
       "        \n",
       "    .jupman-warn {    \n",
       "        background-color: #fcf2f2;        \n",
       "        border-left: 5px solid #dfb5b4;\n",
       "        padding: 1.5em;\n",
       "        font-weight:bold;\n",
       "        margin-bottom:10px;\n",
       "    }\n",
       "\n",
       "    .jupman-question {    \n",
       "        background-color: rgb(251, 231, 250);\n",
       "        border-left: 5px solid rgb(189, 151, 184);\n",
       "        padding: 1.5em;\n",
       "        font-weight:bold;\n",
       "        margin-bottom:10px;\n",
       "    }\n",
       "        \n",
       "    \n",
       "    .jupman-important {\n",
       "        background-color: rgba(255, 238, 0, 0.31);\n",
       "        border-left: 5px solid rgba(205, 208, 0, 1.0);\n",
       "        color: #000;        \n",
       "        padding: 1.5em;\n",
       "        font-weight:bold;\n",
       "        margin-bottom:10px;\n",
       "    }\n",
       "    \n",
       "    \n",
       "    #jupman-toc ul {                                                                                                                                                                           \n",
       "        list-style-type: none;                                                                                                                                                                 \n",
       "        margin-left: 5px;\n",
       "    }\n",
       "    \n",
       "    #jupman-toc li {        \n",
       "         padding-left: 3px;\n",
       "    }    \n",
       "    \n",
       "    #jupman-toc .h1 {\n",
       "        font-size:110%; \n",
       "        \n",
       "    }\n",
       "    \n",
       "    #jupman-toc .h2 {\n",
       "        font-size:100%; \n",
       "        padding-left:15px;\n",
       "    }\n",
       "\n",
       "    #jupman-toc .h3 {\n",
       "        font-size:100%; \n",
       "        padding-left:25px;\n",
       "    }\n",
       "    \n",
       "    \n",
       "    \n",
       "    #notebook-container {\n",
       "        box-shadow : none;\n",
       "\t-web-kit-box-shadow: none;\n",
       "    }\n",
       "    \n",
       "\n",
       "    #jupman-toc .toc-active {\n",
       "        background: rgba(168, 254, 255, 0.32);\n",
       "    }\n",
       " </style>\n",
       "\n",
       "<script>\n",
       "\n",
       "/*!\n",
       " * toc - jQuery Table of Contents Plugin\n",
       " * v0.3.2\n",
       " * http://projects.jga.me/toc/\n",
       " * copyright Greg Allen 2014\n",
       " * MIT License\n",
       "*/\n",
       "!function(a){a.fn.smoothScroller=function(b){b=a.extend({},a.fn.smoothScroller.defaults,b);var c=a(this);return a(b.scrollEl).animate({scrollTop:c.offset().top-a(b.scrollEl).offset().top-b.offset},b.speed,b.ease,function(){var a=c.attr(\"id\");a.length&&(history.pushState?history.pushState(null,null,\"#\"+a):document.location.hash=a),c.trigger(\"smoothScrollerComplete\")}),this},a.fn.smoothScroller.defaults={speed:400,ease:\"swing\",scrollEl:\"body,html\",offset:0},a(\"body\").on(\"click\",\"[data-smoothscroller]\",function(b){b.preventDefault();var c=a(this).attr(\"href\");0===c.indexOf(\"#\")&&a(c).smoothScroller()})}(jQuery),function(a){var b={};a.fn.toc=function(b){var c,d=this,e=a.extend({},jQuery.fn.toc.defaults,b),f=a(e.container),g=a(e.selectors,f),h=[],i=e.activeClass,j=function(b,c){if(e.smoothScrolling&&\"function\"==typeof e.smoothScrolling){b.preventDefault();var f=a(b.target).attr(\"href\");e.smoothScrolling(f,e,c)}a(\"li\",d).removeClass(i),a(b.target).parent().addClass(i)},k=function(){c&&clearTimeout(c),c=setTimeout(function(){for(var b,c=a(window).scrollTop(),f=Number.MAX_VALUE,g=0,j=0,k=h.length;k>j;j++){var l=Math.abs(h[j]-c);f>l&&(g=j,f=l)}a(\"li\",d).removeClass(i),b=a(\"li:eq(\"+g+\")\",d).addClass(i),e.onHighlight(b)},50)};return e.highlightOnScroll&&(a(window).bind(\"scroll\",k),k()),this.each(function(){var b=a(this),c=a(e.listType);g.each(function(d,f){var g=a(f);h.push(g.offset().top-e.highlightOffset);var i=e.anchorName(d,f,e.prefix);if(f.id!==i){a(\"<span/>\").attr(\"id\",i).insertBefore(g)}var l=a(\"<a/>\").text(e.headerText(d,f,g)).attr(\"href\",\"#\"+i).bind(\"click\",function(c){a(window).unbind(\"scroll\",k),j(c,function(){a(window).bind(\"scroll\",k)}),b.trigger(\"selected\",a(this).attr(\"href\"))}),m=a(\"<li/>\").addClass(e.itemClass(d,f,g,e.prefix)).append(l);c.append(m)}),b.html(c)})},jQuery.fn.toc.defaults={container:\"body\",listType:\"<ul/>\",selectors:\"h1,h2,h3\",smoothScrolling:function(b,c,d){a(b).smoothScroller({offset:c.scrollToOffset}).on(\"smoothScrollerComplete\",function(){d()})},scrollToOffset:0,prefix:\"toc\",activeClass:\"toc-active\",onHighlight:function(){},highlightOnScroll:!0,highlightOffset:100,anchorName:function(c,d,e){if(d.id.length)return d.id;var f=a(d).text().replace(/[^a-z0-9]/gi,\" \").replace(/\\s+/g,\"-\").toLowerCase();if(b[f]){for(var g=2;b[f+g];)g++;f=f+\"-\"+g}return b[f]=!0,e+\"-\"+f},headerText:function(a,b,c){return c.text()},itemClass:function(a,b,c,d){return d+\"-\"+c[0].tagName.toLowerCase()}}}(jQuery);\n",
       "\n",
       "/*\n",
       " * JUPYTER MANAGER JavaScript https://github.com/DavidLeoni/jupman \n",
       " * \n",
       " */\n",
       "\n",
       " function toggleVisibility(what){\n",
       "    var e = document.getElementById(what);\n",
       "    if(e.style.display == 'block')\n",
       "      e.style.display = 'none';\n",
       "    else\n",
       "      e.style.display = 'block';\n",
       "  }\n",
       "    \n",
       " \n",
       "\n",
       "\n",
       "function showthis(url) {\n",
       "    window.open(url, \"pres\", \"toolbar=yes,scrollbars=yes,resizable=yes,top=10,left=400,width=500,height=500\");\n",
       "    return(false);\n",
       "}\n",
       "\n",
       "var jupman = {\n",
       "    \n",
       "    isReduced : function(){\n",
       "        return $(window).width() < 924;\n",
       "    },\n",
       "    hoverToc : function(){\n",
       "        return $('#jupman-toc:hover').length != 0;\n",
       "    },\n",
       "    resize : function(){\n",
       "        if (jupman.isReduced()){\n",
       "            $(\"#jupman-toc\").hide();\n",
       "        } else {\n",
       "            $(\"#jupman-toc\").show();\n",
       "            $(\"#jupman-toc\").css(\"background\",\"rgba(255, 255, 255, 0)\");\n",
       "            \n",
       "            tocParams = {\n",
       "                \n",
       "                'selectors': 'h1,h2,h3', //elements to use as headings\n",
       "                'container': 'body', //element to find all selectors in\n",
       "                'smoothScrolling': true, //enable or disable smooth scrolling on click\n",
       "                 //doesn't work  'prefix': 'jupman-toc', //prefix for anchor tags and class names\n",
       "                //'onHighlight': function(el) {}, //called when a new section is highlighted \n",
       "                'highlightOnScroll': true, //add class to heading that is currently in focus\n",
       "                'highlightOffset': 100, //offset to trigger the next headline\n",
       "                'anchorName': function(i, heading, prefix) { //custom function for anchor name\n",
       "                    return prefix+i;\n",
       "                },\n",
       "                'headerText': function(i, heading, $heading) { //custom function building the header-item text                  \n",
       "                    return $heading.text().replace(\"Â¶\",\"\");\n",
       "                },\n",
       "                'itemClass': function(i, heading, $heading, prefix) { // custom function for item class\n",
       "                    return $heading[0].tagName.toLowerCase();\n",
       "                }                \n",
       "            }\n",
       "            \n",
       "            $('#jupman-toc').toc(tocParams);\n",
       "        }\n",
       "    },\n",
       "    /**\n",
       "        Hides cell input\n",
       "        \n",
       "        Renamed in 0.19\n",
       "    */    \n",
       "    hideCell : function(prefixOrRegex){\n",
       "        \n",
       "        $('.border-box-sizing .code_cell pre').filter(function() { \n",
       "                var t = $(this).text();\n",
       "                if (typeof prefixOrRegex == \"string\"){\n",
       "                    return t.indexOf(prefixOrRegex) === 0;\n",
       "                } else if ( prefixOrRegex instanceof RegExp){\n",
       "                    return t.match(prefixOrRegex);\n",
       "                } else {\n",
       "                    console.error(\"Invalid argument:\", prefixOrRegex);\n",
       "                    throw new Error(\"Invalid argument!\");\n",
       "                }\n",
       "            }).parents('div .cell .input').hide();        \n",
       "    },\n",
       "    /**\n",
       "        Hides the cell input AND output.\n",
       "        \n",
       "        @since 0.19\n",
       "    */\n",
       "    hideCellAll : function(prefix){        \n",
       "        $('.border-box-sizing .code_cell pre').filter(function() { \n",
       "                return $(this).text().indexOf(prefix) === 0; \n",
       "            }).parents('div .cell ').hide();        \n",
       "    },\n",
       "    \n",
       "    init : function(){\n",
       "\n",
       "       var toc = $(\"<div>\").attr(\"id\", \"jupman-toc\");              \n",
       "       var indexLink = $(\"<a>\")\n",
       "                        .addClass(\"jupman-nav-item\")\n",
       "                        .attr(\"href\",\"index.html#Chapters\")\n",
       "                        .text(\"jupman\");\n",
       "       \n",
       "       \n",
       "       \n",
       "       var candidateTitleText = $(\".jupman-title\").text();              \n",
       "                                  \n",
       "                    \n",
       "       \n",
       "       var nav = $(\"<div>\")\n",
       "                     .attr(\"id\", \"jupman-nav\")\n",
       "                    .append(indexLink);       \n",
       "       \n",
       "       if (candidateTitleText.length !== 0){\n",
       "\n",
       "           var title = $(\"<span>\")\n",
       "                    .addClass(\"jupman-nav-item\")\n",
       "                    .css(\"padding-left\",\"8px\")\n",
       "                    .text(candidateTitleText);\n",
       "            nav.append(\"<br>\")\n",
       "                .append(\"<br>\")\n",
       "                .append(title);                                \n",
       "        }\n",
       "                     \n",
       "       // TODO THIS HIDE STUFF DOES NOT WORK ANYMORE AFTER PORTING TO NBSPHINX \n",
       "       // INSTEAD, YOU SHOULD EDIT \n",
       "       jupman.hideCell(\"%%HTML\");\n",
       "       jupman.hideCell(\"import jupman\");   \n",
       "       \n",
       "        // TODO this is a bit too hacky   \n",
       "       jupman.hideCell(/from exercise(.+)_solution import \\*/)\n",
       "       \n",
       "       jupman.hideCell(\"jupman.init()\"); \n",
       "       jupman.hideCell(\"jupman.show_run(\");\n",
       "       jupman.hideCell(\"nxpd.draw(\");\n",
       "       jupman.hideCellAll(\"jupman.run(\"); \n",
       "              \n",
       "       if ($(\"#jupman-toc\").length === 0){\n",
       "           $(\"body\").append(toc);       \n",
       "       } else {\n",
       "           $(\"#jupman-toc\").replaceWith(toc);\n",
       "       }\n",
       "       \n",
       "       if ($(\"#jupman-nav\").length === 0){\n",
       "           $(\"body\").append(nav);       \n",
       "       } else {\n",
       "           $(\"#jupman-nav\").replaceWith(nav);\n",
       "       }              \n",
       "       \n",
       "       jupman.resize();\n",
       "    }\n",
       "}\n",
       "\n",
       "$( window ).resize(function() {\n",
       "    jupman.resize();\n",
       "});\n",
       "\n",
       "\n",
       "$(\"body\").on(\"mousemove\",function(event) {\n",
       "    if (jupman.isReduced()){\n",
       "        if (event.pageX < 50) {            \n",
       "             $(\"#jupman-toc\").show(); \n",
       "            $(\"#jupman-toc\").css(\"background\",\"rgba(255, 255, 255, 1.0)\");\n",
       "        } else {\n",
       "            \n",
       "            if (jupman.hoverToc()) {                    \n",
       "            } else {\n",
       "                $(\"#jupman-toc\").hide();                        \n",
       "            }\n",
       "                     \n",
       "/*            if ($(\"#jupman-toc\").is(\":visible\")){\n",
       "                if (jupman.hoverToc()) {                    \n",
       "                } else {\n",
       "                    $(\"#jupman-toc\").hide();                        \n",
       "                }\n",
       "            } else {\n",
       "                if (jupman.hoverToc())\n",
       "                  show\n",
       "              } else {\n",
       "                $(\"#jupman-toc\").hide();                        \n",
       "               }                 \n",
       "            }\n",
       "  */     \n",
       "        }\n",
       "    }\n",
       "});\n",
       "\n",
       "$(document).ready(jupman.init);\n",
       "\n",
       "</script>\n"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import jupman\n",
    "jupman.init()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "# Practical 10\n",
    "\n",
    "In this practical we will will get acquainted with Biopython.\n",
    "\n",
    "## Slides\n",
    "\n",
    "The slides of the introduction can be found here: [Intro](docs/Practical10.pdf)\n",
    "\n",
    "# Biopython\n",
    "\n",
    "From the Biopython tutorial: The Biopython Project is an international association of developers of freely available Python tools for computational molecular biology. The goal of Biopython is to make it as easy as possible to use Python for bioinformatics by creating high-quality, reusable modules and classes. Biopython features include parsers for various Bioinformatics file formats (FASTA, FASTQ, BLAST, PDB, Clustalw, Genbank,...), access to online services (NCBI, Expasy,...), interfaces to common and not-so-common programs (Clustalw, DSSP, MSMS...), a standard sequence class, various clustering modules, a KD tree data structure etc. and even documentation.\n",
    "\n",
    "In this practical we will see some features of Biopython but refer to [biopython documentation](http://biopython.org/wiki/Documentation) to discover all its features, recipes etc.\n",
    "\n",
    "These notes are largely based on what available [here](http://biopython.org/DIST/docs/tutorial/Tutorial.pdf).\n",
    "\n",
    "## Installation of Biopython\n",
    "\n",
    "To test if Biopython is installed in your system you can import the library with:\n",
    "\n",
    "```\n",
    "import Bio\n",
    "```\n",
    "\n",
    "if the execution gives:\n",
    "\n",
    "![](img/pract10/import.png)\n",
    "\n",
    "Biopython needs to be installed.\n",
    "\n",
    "**In windows** installing Biopython should be as easy as opening the command prompt as admininstrator (typing ```cmd``` and then right clicking on the link choosing run as admininstrator) and then ```pip3 install biopython```.\n",
    "\n",
    "**In linux**  ```sudo pip3 install biopython``` will install biopython for python3 up to python3.5. On python 3.6, the command is: ```python3.6 -m pip install biopython```.\n",
    "\n",
    "\n",
    "Detailed installation instructions can be found [here](http://biopython.org/DIST/docs/install/Installation.html).\n",
    "\n",
    "## General overview\n",
    "\n",
    "Biopython provides the capability of parsing standard bioinformatics formats into python usable data structures. Some of the supported tools and formats are:\n",
    "\n",
    "1. Blast output â both from standalone and WWW Blast\n",
    "2. Clustalw\n",
    "3. FASTA\n",
    "4. GenBank\n",
    "5. PubMed and Medline\n",
    "6. ExPASy files, like Enzyme and Prosite\n",
    "7. SCOP, including âdomâ and âlinâ files\n",
    "8. UniGene\n",
    "9. SwissProt\n",
    "\n",
    "among the others. In particular it provides a sequence class to perform the most common operations on sequences, parsing of alignments, clustering etc. and it can quickly connect to the most popular databases to download data or parse information on the fly.\n",
    "\n",
    "\n",
    "## Sequence objects\n",
    "\n",
    "Biopython has a specific object ```Seq``` to deal with biological sequences which is a lot more powerful than the ```str``` object of python when it comes to manage biological data. The object **Seq** has methods like **translate()** and **reverse_complement()** which are very handy.  \n",
    "\n",
    "As in the case of python's ```str``` objects, the object ```Seq``` is immutable. The mutable version is ```MutableSeq```.\n",
    "\n",
    "All the information on Seq objects can be found [here](http://biopython.org/DIST/docs/api/Bio.Seq.Seq-class.html).\n",
    "\n",
    "### The alphabet\n",
    "\n",
    "```Seq``` objects are composed by two information: the actual sequence and the alphabet the sequence refers to. We normally will use the [IUPAC](http://www.chem.qmw.ac.uk/iupac/) alphabet as defined in the ```Bio.Alphabet.IUPAC``` which provides a suitable alphabet to deal with DNA, RNA and proteins. Several options are available, including ```Bio.Alphabet.generic_dna``` or ```Bio.Alphabet.generic_protein```, ```Bio.Alphabet.ThreeLetterProtein``` and other.\n",
    "\n",
    "Specifying an alphabet quickly tells you what information you are dealing with and moreover it provides a way to specify a constraint on the data.\n",
    "\n",
    "If the definition of the alphabet is not specified, the default one is used. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "S: GATTACATAATA\n",
      "S's alphabet: Alphabet()\n",
      "dna_seq: GATTATACGTAC\n",
      "dna_seq's alphabet: IUPACUnambiguousDNA()\n",
      "my_prot: MGNAAAAKKGSEQE\n",
      "my_prot's alphabet: IUPACProtein()\n"
     ]
    }
   ],
   "source": [
    "from Bio.Seq import Seq\n",
    "from Bio.Alphabet import IUPAC\n",
    "\n",
    "#No alphabet specified\n",
    "s = Seq(\"GATTACATAATA\")\n",
    "dna_seq = Seq(\"GATTATACGTAC\", IUPAC.unambiguous_dna)\n",
    "print(\"S:\", s)\n",
    "print(\"S's alphabet:\", s.alphabet)\n",
    "print(\"dna_seq:\", dna_seq)\n",
    "print(\"dna_seq's alphabet:\", dna_seq.alphabet)\n",
    "\n",
    "my_prot = Seq(\"MGNAAAAKKGSEQE\", IUPAC.protein)\n",
    "print(\"my_prot:\", my_prot)\n",
    "print(\"my_prot's alphabet:\", my_prot.alphabet)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Looping, slicing and concatenation\n",
    "\n",
    "Note also that **Seq objects behave like strings**, but the consistency of the alphabet is checked too. For example we cannot concatenate a **unambiguous_dna** with a **IUPAC.protein** sequence.  This is the error message we would get:\n",
    "\n",
    "![](img/pract10/mess.png)\n",
    "\n",
    "To do it one should change the alphabet to ```generic_alphabet```:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GATTATACGTACMGNAAAAKKGSEQE\n"
     ]
    }
   ],
   "source": [
    "from Bio.Seq import Seq\n",
    "from Bio.Alphabet import generic_alphabet\n",
    "\n",
    "dna_seq = Seq(\"GATTATACGTAC\", IUPAC.unambiguous_dna)\n",
    "my_prot = Seq(\"MGNAAAAKKGSEQE\", IUPAC.protein)\n",
    "\n",
    "my_prot.alphabet = generic_alphabet\n",
    "\n",
    "#Does it really make sense though?!?\n",
    "print(dna_seq + my_prot)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can also loop through the elements and slice strings with the usual [S:E:step] operator (and as usual S is included, E is excluded). \n",
    "We can also convert a Seq object into a string."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "G A T T A T A C G T A C G G C T A \n",
      "ATACGT\n",
      "Reversed:  ATCGGCATGCATATTAG\n",
      "As string: GATTATACGTACGGCTA\n",
      "<class 'str'>\n"
     ]
    }
   ],
   "source": [
    "from Bio.Seq import Seq\n",
    "from Bio.Alphabet import IUPAC\n",
    "\n",
    "dna_seq = Seq(\"GATTATACGTACGGCTA\", IUPAC.unambiguous_dna)\n",
    "\n",
    "for base in dna_seq:\n",
    "    print(base, end = \" \")\n",
    "    \n",
    "print(\"\")\n",
    "\n",
    "sub_seq = dna_seq[4:10]\n",
    "print(sub_seq)\n",
    "\n",
    "#Let's reverse the string:\n",
    "\n",
    "print(\"Reversed: \", dna_seq[::-1])\n",
    "#from Seq to string:\n",
    "dna_str = str(dna_seq)\n",
    "print(\"As string:\", dna_str)\n",
    "print(type(dna_str))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The object Seq provides several methods similar to the corresponding methods of strings:\n",
    "\n",
    "1. ```Seq.count(s)``` : counts the number of times s appears in the sequence;\n",
    "2. ```Seq.upper()``` : makes the sequence of the object Seq upper case\n",
    "3. ```Seq.lower()``` : makes the sequence of the object Seq lower case\n",
    "\n",
    "**Example:**\n",
    "Let's compute the GC content of the sequence \"GATTRWWACGTACGGCTASATTACSCCGGCTA\"."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GC % is 40.62\n",
      "GC % is 46.88\n",
      "original: GATTRWWACGTACGGCTASATTACSCCGGCTA\n",
      "lower_case: gattrwwacgtacggctasattacsccggcta\n"
     ]
    }
   ],
   "source": [
    "from Bio.Seq import Seq\n",
    "from Bio.Alphabet import IUPAC\n",
    "from Bio.SeqUtils import GC\n",
    "\n",
    "dna_seq = Seq(\"GATTRWWACGTACGGCTASATTACSCCGGCTA\", IUPAC.ambiguous_dna)\n",
    "\n",
    "gc = (dna_seq.count(\"G\") + dna_seq.count(\"C\"))/len(dna_seq) \n",
    "print(\"GC % is {:.2f}\".format(100*gc))\n",
    "\n",
    "correct_gc = GC(dna_seq)\n",
    "print(\"GC % is {:.2f}\".format(correct_gc))\n",
    "\n",
    "print(\"original:\", dna_seq)\n",
    "print(\"lower_case:\",dna_seq.lower())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The different is that Biopython's GC method takes into account ambiguous nucleotides like (S which stands for C or G [ref iupac](https://www.bioinformatics.org/sms/iupac.html)). This is a method provided in the **SeqUtils** module (more info on all the methods [here](http://biopython.org/DIST/docs/api/Bio.SeqUtils-module.html))."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Complement and reverse complement\n",
    "\n",
    "A usual operation on DNA strings is the reverse complement. Biopython provides two specific methods to do that:\n",
    "\n",
    "1. ```Seq.complement()``` to complement the sequence\n",
    "2. ```Seq.reverse_complement()``` to reverse complement the sequence.\n",
    "\n",
    "Note that these methods will fail with a ```Value Error``` if we try to apply them to something that is not a nucleotide sequence (e.g. proteins)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sequence (alphabet: IUPACUnambiguousDNA())\n",
      "GATCGATGGGCCTATATAGGATCGAAAATCGC\n",
      "\n",
      "Complement (alphabet: IUPACUnambiguousDNA())\n",
      "CTAGCTACCCGGATATATCCTAGCTTTTAGCG\n",
      "\n",
      "Rev Complement (alphabet: IUPACUnambiguousDNA())\n",
      "GCGATTTTCGATCCTATATAGGCCCATCGATC\n"
     ]
    }
   ],
   "source": [
    "from Bio.Seq import Seq\n",
    "from Bio.Alphabet import IUPAC\n",
    "\n",
    "my_seq = Seq(\"GATCGATGGGCCTATATAGGATCGAAAATCGC\", IUPAC.unambiguous_dna)\n",
    "\n",
    "print(\"Sequence (alphabet: {})\".format(my_seq.alphabet))\n",
    "print(my_seq)\n",
    "comp  = my_seq.complement()\n",
    "print(\"\")\n",
    "print(\"Complement (alphabet: {})\".format(comp.alphabet))\n",
    "print(comp)\n",
    "print(\"\")\n",
    "revcomp = my_seq.reverse_complement()\n",
    "print(\"Rev Complement (alphabet: {})\".format(revcomp.alphabet))\n",
    "print(revcomp)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Transcription and translation\n",
    "\n",
    "Remember that:\n",
    "\n",
    "![](img/pract10/trans_transl.png)\n",
    "\n",
    "[taken from the Biopython tutorial].\n",
    "\n",
    "\n",
    "Biopython has a method to transcribe a DNA sequence into a mRNA sequence (and one to go back):\n",
    "\n",
    "1. ```Seq.transcribe()``` that transcribes a DNA sequence into mRNA (converting Ts in Us).\n",
    "2. ```Seq.back_transcribe()``` that back-transcribes a mRNA sequence into DNA (converting Us in Ts).\n",
    "\n",
    "The alphabet of the transcribed sequence is converted into mrna. Note that transcribing a protein will result in a \n",
    "**ValueError**.\n",
    "\n",
    "Alike, a DNA or mRNA sequence can be translated into a protein with:\n",
    "\n",
    "```Seq.translate(table=N, to_stop=True/False, stop_symbol =symbol)``` that converts a DNA sequence or mRNA into the corresponding protein. \n",
    "\n",
    "Optionally, the translate method accepts a parameter ```table``` to specify the NCBI translation table used (see below for codes). ```to_stop``` makes the translation stop at the first stop codon if True (**in this case the stop codon is not translated!**), ```stop_symbol``` to replace the asterisk with any other stop character (Es. \"@\").\n",
    "\n",
    "![](img/pract10/ncbi_table.png)\n",
    "Stop codons will be represented by an asterisk. \n",
    "\n",
    "**Example:**\n",
    "Let's translate into mRNA the coding sequence: ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG\n",
      "AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG\n",
      "\n",
      "IUPACUnambiguousRNA()\n",
      "\n",
      "... and back\n",
      "ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG\n",
      "\n",
      "Translation to protein:\n",
      "MAIVMGR*KGAR*\n",
      "\n",
      "Up to first stop:\n",
      "MAIVMGR\n",
      "\n",
      "Mitocondrial translation: (TGA is W!)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Seq('MAIVMGRWKGAR*', HasStopCodon(IUPACProtein(), '*'))"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from Bio.Seq import Seq\n",
    "from Bio.Alphabet import IUPAC\n",
    "\n",
    "coding_dna = Seq(\"ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG\", IUPAC.unambiguous_dna)\n",
    "print(coding_dna)\n",
    "\n",
    "mrna = coding_dna.transcribe()\n",
    "print(mrna)\n",
    "print(\"\")\n",
    "print(mrna.alphabet)\n",
    "print(\"\")\n",
    "print(\"... and back\")\n",
    "print(mrna.back_transcribe())\n",
    "print(\"\")\n",
    "print(\"Translation to protein:\")\n",
    "prot = mrna.translate()\n",
    "print(prot)\n",
    "print(\"\")\n",
    "print(\"Up to first stop:\")\n",
    "print(mrna.translate(to_stop = True))\n",
    "print(\"\")\n",
    "print(\"Mitocondrial translation: (TGA is W!)\")\n",
    "mit_prot = mrna.translate(table=2)\n",
    "mit_prot\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Translation tables can be accessed runtime through:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Table 1 Standard, SGC0\n",
      "\n",
      "  |  T      |  C      |  A      |  G      |\n",
      "--+---------+---------+---------+---------+--\n",
      "T | TTT F   | TCT S   | TAT Y   | TGT C   | T\n",
      "T | TTC F   | TCC S   | TAC Y   | TGC C   | C\n",
      "T | TTA L   | TCA S   | TAA Stop| TGA Stop| A\n",
      "T | TTG L(s)| TCG S   | TAG Stop| TGG W   | G\n",
      "--+---------+---------+---------+---------+--\n",
      "C | CTT L   | CCT P   | CAT H   | CGT R   | T\n",
      "C | CTC L   | CCC P   | CAC H   | CGC R   | C\n",
      "C | CTA L   | CCA P   | CAA Q   | CGA R   | A\n",
      "C | CTG L(s)| CCG P   | CAG Q   | CGG R   | G\n",
      "--+---------+---------+---------+---------+--\n",
      "A | ATT I   | ACT T   | AAT N   | AGT S   | T\n",
      "A | ATC I   | ACC T   | AAC N   | AGC S   | C\n",
      "A | ATA I   | ACA T   | AAA K   | AGA R   | A\n",
      "A | ATG M(s)| ACG T   | AAG K   | AGG R   | G\n",
      "--+---------+---------+---------+---------+--\n",
      "G | GTT V   | GCT A   | GAT D   | GGT G   | T\n",
      "G | GTC V   | GCC A   | GAC D   | GGC G   | C\n",
      "G | GTA V   | GCA A   | GAA E   | GGA G   | A\n",
      "G | GTG V   | GCG A   | GAG E   | GGG G   | G\n",
      "--+---------+---------+---------+---------+--\n",
      "Table 2 Vertebrate Mitochondrial, SGC1\n",
      "\n",
      "  |  T      |  C      |  A      |  G      |\n",
      "--+---------+---------+---------+---------+--\n",
      "T | TTT F   | TCT S   | TAT Y   | TGT C   | T\n",
      "T | TTC F   | TCC S   | TAC Y   | TGC C   | C\n",
      "T | TTA L   | TCA S   | TAA Stop| TGA W   | A\n",
      "T | TTG L   | TCG S   | TAG Stop| TGG W   | G\n",
      "--+---------+---------+---------+---------+--\n",
      "C | CTT L   | CCT P   | CAT H   | CGT R   | T\n",
      "C | CTC L   | CCC P   | CAC H   | CGC R   | C\n",
      "C | CTA L   | CCA P   | CAA Q   | CGA R   | A\n",
      "C | CTG L   | CCG P   | CAG Q   | CGG R   | G\n",
      "--+---------+---------+---------+---------+--\n",
      "A | ATT I(s)| ACT T   | AAT N   | AGT S   | T\n",
      "A | ATC I(s)| ACC T   | AAC N   | AGC S   | C\n",
      "A | ATA M(s)| ACA T   | AAA K   | AGA Stop| A\n",
      "A | ATG M(s)| ACG T   | AAG K   | AGG Stop| G\n",
      "--+---------+---------+---------+---------+--\n",
      "G | GTT V   | GCT A   | GAT D   | GGT G   | T\n",
      "G | GTC V   | GCC A   | GAC D   | GGC G   | C\n",
      "G | GTA V   | GCA A   | GAA E   | GGA G   | A\n",
      "G | GTG V(s)| GCG A   | GAG E   | GGG G   | G\n",
      "--+---------+---------+---------+---------+--\n"
     ]
    }
   ],
   "source": [
    "from Bio.Data import CodonTable\n",
    "#equivalent to:\n",
    "#standard_table = CodonTable.unambiguous_dna_by_id[1]\n",
    "standard_table = CodonTable.unambiguous_dna_by_name[\"Standard\"]\n",
    "#equivalent to:\n",
    "#standard_table = CodonTable.unambiguous_dna_by_id[2]\n",
    "mito_table = CodonTable.unambiguous_dna_by_name[\"Vertebrate Mitochondrial\"]\n",
    "\n",
    "\n",
    "print(standard_table)\n",
    "\n",
    "print(mito_table)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### MutableSeq\n",
    "\n",
    "Seq objects are **immutable and therefore we cannot change their value**. If we need to do so, we need to convert them into ```MutableSeq``` objects to avoid a runtime **TypeError** that will occur when we try to change a Seq. Seq to MutableSeq conversion can be done by:\n",
    "\n",
    "1. ```Seq.tomutable()``` to convert a Seq to a MutableSeq.\n",
    "2. ```MutableSeq.toseq()``` to convert a MutableSeq to a Seq.\n",
    "\n",
    "This will allow us to assign the value of characters within the MutableSeq with ```MutableSeq[ind] = newValue```. Note also that methods like complement change the actual object if applied to a MutableSeq:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG\n",
      "ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG\n",
      "ATGGCCATTGTATATATATACTGAAAGGGTGCCCGATAG\n",
      "\n",
      "Rev comp (Seq not changed!):\n",
      "ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG\n",
      "Rev comp (MutableSeq changed!):\n",
      "CTATCGGGCACCCTTTCAGTATATATATACAATGGCCAT\n"
     ]
    }
   ],
   "source": [
    "from Bio.Seq import Seq\n",
    "from Bio.Alphabet import IUPAC\n",
    "\n",
    "coding_dna = Seq(\"ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG\", IUPAC.unambiguous_dna)\n",
    "print(coding_dna)\n",
    "mut_dna = coding_dna.tomutable()\n",
    "print(mut_dna)\n",
    "mut_dna[10:20] = \"TATATATATA\"\n",
    "print(mut_dna)\n",
    "print(\"\")\n",
    "print(\"Rev comp (Seq not changed!):\")\n",
    "coding_dna.reverse_complement()\n",
    "print(coding_dna)\n",
    "print(\"Rev comp (MutableSeq changed!):\")\n",
    "mut_dna.reverse_complement()\n",
    "print(mut_dna)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Sequence annotation\n",
    "\n",
    "Sequence records are annotations associated to sequences. In biopython there is a ```SeqRecord``` object to store this information. ```SeqRecord``` are part of the ```Bio.SeqRecord``` module which therefore needs to be imported prior to be used.\n",
    "\n",
    "The **SeqRecord** class has the following attributes:\n",
    "\n",
    "1. ```SeqRecord.seq``` : the sequence (the Seq object)\n",
    "2. ```SeqRecord.id``` : the identifier of the sequence, typically an accession number\n",
    "3. ```SeqRecord.name``` : a \"common\" name or identifier sometimes identical to the accession number\n",
    "4. ```SeqRecord.description``` : a human readable description of the sequence\n",
    "5. ```SeqRecord.letter_annotations``` : a per letter annotation using a restricted dictionary\n",
    "6. ```SeqRecord.annotations``` : a dictionary of unstructured annotation\n",
    "7. ```SeqRecord.features``` : a list of SeqFeature objects with more structured information (e.g. genes pos). \n",
    "8. ```SeqRecord.dbxrefs``` : a list of database cross references.\n",
    "\n",
    "Information on SeqFeature objects can be found [here](http://biopython.org/DIST/docs/tutorial/Tutorial.html#htoc38).\n",
    "\n",
    "Although you can manually create a SeqRecord, normally you will access SeqRecords populated by methods like SeqIO, [here](http://biopython.org/DIST/docs/tutorial/Tutorial.html#htoc33) you can find how to manually create a SeqRecord.\n",
    "\n",
    "**Example:** \n",
    "\n",
    "Read a fasta file [NC005816.fna](file_samples/NC_005816.fna) containing the whole sequence for Yersinia pestis biovar Microtus str. 91001 plasmid pPCP1 and retrieve some information about the sequence."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ID: gi|45478711|ref|NC_005816.1|\n",
      "Name: gi|45478711|ref|NC_005816.1|\n",
      "Description: gi|45478711|ref|NC_005816.1| Yersinia pestis biovar Microtus str. 91001 plasmid pPCP1, complete sequence\n",
      "Number of features: 0\n",
      "Seq('TGTAACGAACGGTGCAATAGTGATCCACACCCAACGCCTGAAATCAGATCCAGG...CTG', SingleLetterAlphabet())\n",
      "\n",
      "Sequence [first 30 bases]:\n",
      "TGTAACGAACGGTGCAATAGTGATCCACAC\n",
      "\n",
      "The id:\n",
      "gi|45478711|ref|NC_005816.1|\n",
      "\n",
      "The description:\n",
      "gi|45478711|ref|NC_005816.1| Yersinia pestis biovar Microtus str. 91001 plasmid pPCP1, complete sequence\n",
      "\n",
      "The record is a:  <class 'Bio.SeqRecord.SeqRecord'>\n"
     ]
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "\n",
    "record = SeqIO.read(\"file_samples/NC_005816.fna\", \"fasta\")\n",
    "\n",
    "print(record)\n",
    "print(\"\")\n",
    "print(\"Sequence [first 30 bases]:\")\n",
    "print(record.seq[0:30])\n",
    "print(\"\")\n",
    "print(\"The id:\")\n",
    "print(record.id)\n",
    "print(\"\")\n",
    "print(\"The description:\")\n",
    "print(record.description)\n",
    "print(\"\")\n",
    "print(\"The record is a: \", type(record))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Sequence I/O\n",
    "\n",
    "The ```Bio.SeqIO``` module aims to provide a simple way to work with several different sequence file formats. Detailed information on the module can be found [here](http://biopython.org/wiki/SeqIO).\n",
    "\n",
    "<div class=\"alert alert-warning\">\n",
    "\n",
    "**WARNING:** \n",
    "When dealing with very large FASTA or FASTQ files, the overhead of working with all these objects can make scripts too slow. In this case ```SimpleFastaParser``` and ```FastqGeneralIterator``` parsers might be better as they which return just a tuple of strings for each record.\n",
    "</div>\n",
    "\n",
    "### BioSeqIO.parse()\n",
    "\n",
    "The method ```Bio.SeqIO.parse``` is used to parse some sequence data into a **SeqRecord iterator**. In particular, the basic syntax is:\n",
    "```\n",
    "SeqRecordIterator = Bio.SeqIO.parse(filename, file_format)\n",
    "```\n",
    "\n",
    "where ```filename``` is typically an open handle to a file and ```file_format``` is a lower case string describing the file format. Possible options include **fasta, fastq-illumina, abi, ace, clustal...** all the options are available [here](http://biopython.org/wiki/SeqIO). \n",
    "\n",
    "**Example:**\n",
    "Let's read the first 10 entries in the .fasta file [contigs82.fasta](files_sample/contigs82.fasta) printing off the length of the sequence and the first 50 bases of each sequence followed by \"...\"."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Seq MDC020656.85 has length 2802\n",
      "GAGGGGTTTAGTTCCTCATACTCGCAAAGCAAAGATACATAAATTTAGAA...\n",
      "\n",
      "Seq MDC001115.177 has length 3118\n",
      "TGAATGGTGAAAATTAGCCAGAAGATCTTCTCCACACATGACATATGCAT...\n",
      "\n",
      "Seq MDC013284.379 has length 5173\n",
      "TATCGTTTCCTCTGAGTAGAATATCGTTATAACAAGATTTTTTTTTTCCT...\n",
      "\n",
      "Seq MDC018185.243 has length 22724\n",
      "CCATTAGTGACCCCCAATGCGGATTAACCAAGCACGGTCAAGATAACCAA...\n",
      "\n",
      "Seq MDC018185.241 has length 23761\n",
      "AAAACGAGGAAAATCCATCTTGATGAACAGGAGATGCGGAGGAAAAAAAT...\n",
      "\n",
      "Seq MDC004527.213 has length 3551\n",
      "CTTGTATGTTGAAGCTTTGTGAGTGGAGCATATAGGTTGAGGTAGTGTTC...\n",
      "\n",
      "Seq MDC003661.174 has length 3334\n",
      "AGTAAGGTTGTAGAATTCAATTTCCACGGAACTCCAGCACAGCTTAGGCA...\n",
      "\n",
      "Seq MDC012176.157 has length 2236\n",
      "ATAGTCACTGACAGCCGGTGACAGTTCGACAACAATAAGTTCAGGATGTT...\n",
      "\n",
      "Seq MDC001204.812 has length 5804\n",
      "TAATGTCCGAAAAAGATGAAAATGTAGTTTGCATGAAGAAGGAGCTGTCC...\n",
      "\n",
      "Seq MDC001204.810 has length 9798\n",
      "ATGCTGCTGACATATTCCAATCGACTACACGTCACATCCGTTGCTATTAT...\n",
      "\n"
     ]
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "cnt = 0\n",
    "for seq_record in SeqIO.parse(\"file_samples/contigs82.fasta\", \"fasta\"):\n",
    "    if(cnt == 10):\n",
    "        break\n",
    "    print(\"Seq {} has length {}\".format(seq_record.id, len(seq_record)))\n",
    "    print(seq_record.seq[:50]+\"...\")\n",
    "    print(\"\")\n",
    "    cnt += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Note that ```Bio.SeqIO.parse``` returns an iterator, therefore it is possible to manually fetch one SeqRecord after the other with the ```next(iterator)``` method.\n",
    "\n",
    "**Example:**\n",
    "Let's get the first 3 entries of the .fasta file [contigs82.fasta](files_sample/contigs82.fasta) printing off the length of the sequence and the first 50 bases of each sequence followed by \"...\"."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1st entry:\n",
      "MDC020656.85  has size  2802\n",
      "GAGGGGTTTAGTTCCTCATACTCGCAAAGCAAAGATACATAAATTTAGAA...\n",
      "\n",
      "2nd entry:\n",
      "MDC001115.177  has size  3118\n",
      "TGAATGGTGAAAATTAGCCAGAAGATCTTCTCCACACATGACATATGCAT...\n",
      "\n",
      "3rd entry:\n",
      "MDC013284.379  has size  5173\n",
      "TATCGTTTCCTCTGAGTAGAATATCGTTATAACAAGATTTTTTTTTTCCT...\n",
      "\n"
     ]
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "\n",
    "seqIterator = SeqIO.parse(\"file_samples/contigs82.fasta\", \"fasta\")\n",
    "\n",
    "labels = [\"1st\",\"2nd\",\"3rd\"]\n",
    "for l in labels:\n",
    "    seqRec = next(seqIterator)\n",
    "    print(l, \"entry:\")\n",
    "    print(seqRec.id, \" has size \", len(seqRec.seq))\n",
    "    print(seqRec.seq[:50]+\"...\")\n",
    "    print(\"\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Sequences as dictionaries\n",
    "\n",
    "The module ```Bio.SeqIO``` also has three different ways to allow random access to elements:\n",
    "\n",
    "1. ```Bio.SeqIO.to_dict(file_handle/iterator)``` : builds a dictionary of all the SeqRecords keeping them in memory and allowing modifications to the records. **This potentially uses a lot of memory but is very fast**;\n",
    "\n",
    "2. ```Bio.SeqIO.index(filename,file_type)``` : builds a sort of read-only dictionary, parses the elements into SeqRecords on demand (i.e. it returns an iterator!). **This method is slower, but more memory efficient**;\n",
    "\n",
    "3. ```Bio.SeqIO.index_db(indexName.idx,filenames, file_format)``` : builds a read-only dictionary, but stores ids and offsets on a SQLite3 database. **It is slower but uses less memory.**\n",
    "\n",
    "\n",
    "The input of ```Bio.SeqIO.to_dict(file_handle/iterator)``` is a open file handler or the iterator to SeqRecords (like the output of Bio.SeqIO.parse()).\n",
    "\n",
    "**Example:**\n",
    "Let's read all the .fasta file [contigs82.fasta](files_sample/contigs82.fasta) in memory printing off the entry corresponding to contig \"MDC019140.399\"."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['MDC018185.243', 'MDC002360.219', 'MDC052568.000', 'MDC020656.85', 'MDC031322.5']\n",
      "\n",
      "Number of sequences:  82\n",
      "\n",
      "Description: MDC019140.399\n",
      "\n",
      "ID: MDC019140.399\n",
      "Name: MDC019140.399\n",
      "Description: MDC019140.399\n",
      "Number of features: 0\n",
      "Seq('GACGGGCGGGGAGGGTTTGGTTTTTTTTGGTTTTTAAAAAATTCAGGTTTNNAA...TTT', SingleLetterAlphabet())\n"
     ]
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "seqDict = SeqIO.to_dict(SeqIO.parse(\"file_samples/contigs82.fasta\", \"fasta\"))\n",
    "\n",
    "print(list(seqDict.keys())[0:5])\n",
    "print(\"\")\n",
    "print(\"Number of sequences: \", len(seqDict))\n",
    "print(\"\")\n",
    "mySeq = seqDict[\"MDC019140.399\"]\n",
    "print(\"Description:\", mySeq.description)\n",
    "print(\"\")\n",
    "print(mySeq)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The input of ```Bio.SeqIO.index(filename,file_type)``` is the file name to read sequences from and the file type (like fasta, genebank,...).\n",
    "\n",
    "**Example:**\n",
    "Let's read all the .fasta file [contigs82.fasta](files_sample/contigs82.fasta) in memory (as an index) printing off the entry corresponding to contig \"MDC019140.399\"."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SeqDict is an iterator:\n",
      "<dict_keyiterator object at 0x7f7f78cc7a48>\n",
      "\n",
      "['MDC018185.243', 'MDC002360.219', 'MDC052568.000', 'MDC020656.85', 'MDC031322.5', 'MDC013723.254', 'MDC008119.414', 'MDC001115.177', 'MDC016296.138', 'MDC000038.355', 'MDC036568.1', 'MDC019140.399', 'MDC017187.318', 'MDC017187.314', 'MDC053310.000', 'MDC002519.240', 'MDC009211.567', 'MDC016278.70', 'MDC014019.318', 'MDC000427.83', 'MDC009211.561', 'MDC018185.242', 'MDC019140.398', 'MDC012865.410', 'MDC020963.162', 'MDC015147.205', 'MDC051637.000', 'MDC017187.311', 'MDC011551.182', 'MDC020963.161', 'MDC003408.117', 'MDC011390.337', 'MDC002838.179', 'MDC010450.877', 'MDC040033.7', 'MDC006346.716', 'MDC018185.241', 'MDC010751.301', 'MDC001204.812', 'MDC000219.190', 'MDC004364.265', 'MDC007154.375', 'MDC018297.229', 'MDC051782.000', 'MDC016621.241', 'MDC017873.233', 'MDC004081.319', 'MDC007995.528', 'MDC002717.156', 'MDC009771.217', 'MDC003995.601', 'MDC019674.147', 'MDC015155.172', 'MDC007838.447', 'MDC026201.7', 'MDC016032.95', 'MDC026961.60', 'MDC012176.157', 'MDC014057.243', 'MDC001802.364', 'MDC021015.302', 'MDC004527.213', 'MDC017187.319', 'MDC013284.379', 'MDC007018.186', 'MDC003661.174', 'MDC022800.298', 'MDC006346.719', 'MDC019067.226', 'MDC010588.505', 'MDC021913.275', 'MDC054294.001', 'MDC021558.159', 'MDC006346.711', 'MDC005174.220', 'MDC024257.15', 'MDC004389.256', 'MDC001204.810', 'MDC013443.168', 'MDC002479.192', 'MDC010588.502', 'MDC007097.457']\n",
      "\n",
      "Number of sequences:  82\n",
      "\n",
      "Description: MDC019140.399\n",
      "\n",
      "ID: MDC019140.399\n",
      "Name: MDC019140.399\n",
      "Description: MDC019140.399\n",
      "Number of features: 0\n",
      "Seq('GACGGGCGGGGAGGGTTTGGTTTTTTTTGGTTTTTAAAAAATTCAGGTTTNNAA...TTT', SingleLetterAlphabet())\n"
     ]
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "seqDict =  SeqIO.index(\"file_samples/contigs82.fasta\", \"fasta\")\n",
    "print(\"SeqDict is an iterator:\")\n",
    "print(seqDict.keys())\n",
    "print(\"\")\n",
    "print(list(seqDict.keys()))\n",
    "print(\"\")\n",
    "print(\"Number of sequences: \", len(seqDict))\n",
    "print(\"\")\n",
    "mySeq = seqDict[\"MDC019140.399\"]\n",
    "print(\"Description:\", mySeq.description)\n",
    "print(\"\")\n",
    "print(mySeq)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The input of ```Bio.SeqIO.index_db(indexName.idx,filenames, file_format)``` is the name of the (SQLite3) index to store (.idx), the file (or files) to index and the format of those files like fasta, genebank,...\n",
    "\n",
    "\n",
    "**Example:**\n",
    "Let's read all the .fasta file [contigs82.fasta](files_sample/contigs82.fasta) in memory (as an index) printing off the entry corresponding to contig \"MDC019140.399\"."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SeqDict is an iterator:\n",
      "<generator object _SQLiteManySeqFilesDict.__iter__ at 0x7f7f78d044c0>\n",
      "\n",
      "['MDC000038.355', 'MDC000219.190', 'MDC000427.83', 'MDC001115.177', 'MDC001204.810', 'MDC001204.812', 'MDC001802.364', 'MDC002360.219', 'MDC002479.192', 'MDC002519.240', 'MDC002717.156', 'MDC002838.179', 'MDC003408.117', 'MDC003661.174', 'MDC003995.601', 'MDC004081.319', 'MDC004364.265', 'MDC004389.256', 'MDC004527.213', 'MDC005174.220', 'MDC006346.711', 'MDC006346.716', 'MDC006346.719', 'MDC007018.186', 'MDC007097.457', 'MDC007154.375', 'MDC007838.447', 'MDC007995.528', 'MDC008119.414', 'MDC009211.561', 'MDC009211.567', 'MDC009771.217', 'MDC010450.877', 'MDC010588.502', 'MDC010588.505', 'MDC010751.301', 'MDC011390.337', 'MDC011551.182', 'MDC012176.157', 'MDC012865.410', 'MDC013284.379', 'MDC013443.168', 'MDC013723.254', 'MDC014019.318', 'MDC014057.243', 'MDC015147.205', 'MDC015155.172', 'MDC016032.95', 'MDC016278.70', 'MDC016296.138', 'MDC016621.241', 'MDC017187.311', 'MDC017187.314', 'MDC017187.318', 'MDC017187.319', 'MDC017873.233', 'MDC018185.241', 'MDC018185.242', 'MDC018185.243', 'MDC018297.229', 'MDC019067.226', 'MDC019140.398', 'MDC019140.399', 'MDC019674.147', 'MDC020656.85', 'MDC020963.161', 'MDC020963.162', 'MDC021015.302', 'MDC021558.159', 'MDC021913.275', 'MDC022800.298', 'MDC024257.15', 'MDC026201.7', 'MDC026961.60', 'MDC031322.5', 'MDC036568.1', 'MDC040033.7', 'MDC051637.000', 'MDC051782.000', 'MDC052568.000', 'MDC053310.000', 'MDC054294.001']\n",
      "\n",
      "Number of sequences:  82\n",
      "\n",
      "Description: MDC019140.399\n",
      "\n",
      "ID: MDC019140.399\n",
      "Name: MDC019140.399\n",
      "Description: MDC019140.399\n",
      "Number of features: 0\n",
      "Seq('GACGGGCGGGGAGGGTTTGGTTTTTTTTGGTTTTTAAAAAATTCAGGTTTNNAA...TTT', SingleLetterAlphabet())\n"
     ]
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "seqDict =  SeqIO.index_db(\"file_samples/ctgs.idx\",\"file_samples/contigs82.fasta\", \"fasta\")\n",
    "print(\"SeqDict is an iterator:\")\n",
    "print(seqDict.keys())\n",
    "print(\"\")\n",
    "print(list(seqDict.keys()))\n",
    "print(\"\")\n",
    "print(\"Number of sequences: \", len(seqDict))\n",
    "print(\"\")\n",
    "mySeq = seqDict[\"MDC019140.399\"]\n",
    "print(\"Description:\", mySeq.description)\n",
    "print(\"\")\n",
    "print(mySeq)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Note that a file ctgs.idx has been created to store ids and offsets. \n",
    "\n",
    "### Writing sequence files\n",
    "\n",
    "SeqRecords can be written out to files by using\n",
    "```\n",
    "N = Bio.SeqIO.write(records,out_filename, file_format)\n",
    "```\n",
    "where **records** is a list of the SeqRecords to write, **out_filename** is the string with the filename to write and **file_format** is the format of the file to write. **N** is the number of sequences written.\n",
    "\n",
    "<div class=\"alert alert-warning\">\n",
    "\n",
    "**WARNING:** \n",
    "If you write a file that is already present, ```SeqIO.write``` will just rewrite it without telling you.\n",
    "</div>\n",
    "\n",
    "**Example:**\n",
    "Write three protein records to file then read the file and print the three sequences. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ID: gi|14150838|gb|AAK54648.1|AF376133_1\n",
      "Name: <unknown name>\n",
      "Description: chalcone synthase [Cucumis sativus]\n",
      "Number of features: 0\n",
      "Seq('MMYQQGCFAGGTVLRLAKDLAENNRGARVLVVCSEITAVTFRGPSETHLDSMVG...SAC', ProteinAlphabet())\n",
      "3 sequences written!\n",
      "\n",
      "Reading file\n",
      "Sequence  1 ( gi|14150838|gb|AAK54648.1|AF376133_1 )\n",
      "MMYQQGCFAGGTVLRLAKDLAENNRGARVLVVCSEITAVTFRGPSETHLDSMVGQALFGDGAGAVIVGSDPDLSVERPLYELVWTGATLLPDSEGAIDGHLREVGLTFHLLKDVPGLISKNIEKSLKEAFTPLGISDWNSTFWIAHPGGPAILDQVEAKLGLKEEKMRATREVLSEYGNMSSAC\n",
      "\n",
      "Sequence  2 ( gi|13919613|gb|AAK33142.1| )\n",
      "YPDYYFRITNREHKAELKEKFQRMCDKSMIKKRYMYLTEEILKENPSMCEYMAPSLDARQDMVVVEIPKLGKEAAVKAIKEWGQ\n",
      "\n",
      "Sequence  3 ( gi|13925890|gb|AAK49457.1| )\n",
      "MVTVEEFRRAQCAEGPATVMAIGTATPSNCVDQSTYPDYYFRITNSEHKVELKEKFKRMCEKSMIKKRYMHLTEEILKENPNICAYMAPSLDARQDIVVVEVPKLGKEAAQKAIKEWGQPKSKITHLVFCTTSGVDMPGCDYQLTKLLGLRPSVKRFMMYQQGCFAGGTVLRMAKDLAENNKGARVLVVCSEITAVTFRGPNDTHLDSLVGQALFGDGAAAVIIGSDPIPEVERPLFELVSAAQTLLPDSEGAIDGHLREVGLTFHLLKDVPGLISKNIEKSLVEAFQPLGISDWNSLFWIAHPGGPAILDQVELKLGLKQEKLKATRKVLSNYGNMSSACVLFILDEMRKASAKEGLGTTGEGLEWGVLFGFGPGLTVETVVLHSVAT\n",
      "\n"
     ]
    }
   ],
   "source": [
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.Alphabet import generic_protein\n",
    "\n",
    "rec1 = SeqRecord(Seq(\"MMYQQGCFAGGTVLRLAKDLAENNRGARVLVVCSEITAVTFRGPSETHLDSMVGQALFGD\" \\\n",
    "+\"GAGAVIVGSDPDLSVERPLYELVWTGATLLPDSEGAIDGHLREVGLTFHLLKDVPGLISK\" \\\n",
    "+\"NIEKSLKEAFTPLGISDWNSTFWIAHPGGPAILDQVEAKLGLKEEKMRATREVLSEYGNM\" \\\n",
    "+\"SSAC\", generic_protein),\n",
    "id=\"gi|14150838|gb|AAK54648.1|AF376133_1\",\n",
    "description=\"chalcone synthase [Cucumis sativus]\")\n",
    "\n",
    "rec2 = SeqRecord(Seq(\"YPDYYFRITNREHKAELKEKFQRMCDKSMIKKRYMYLTEEILKENPSMCEYMAPSLDARQ\" \\\n",
    "+\"DMVVVEIPKLGKEAAVKAIKEWGQ\", generic_protein),\n",
    "id=\"gi|13919613|gb|AAK33142.1|\",\n",
    "description=\"chalcone synthase [Fragaria vesca subsp. bracteata]\")\n",
    "\n",
    "rec3 = SeqRecord(Seq(\"MVTVEEFRRAQCAEGPATVMAIGTATPSNCVDQSTYPDYYFRITNSEHKVELKEKFKRMC\" \\\n",
    "+\"EKSMIKKRYMHLTEEILKENPNICAYMAPSLDARQDIVVVEVPKLGKEAAQKAIKEWGQP\" \\\n",
    "+\"KSKITHLVFCTTSGVDMPGCDYQLTKLLGLRPSVKRFMMYQQGCFAGGTVLRMAKDLAEN\" \\\n",
    "+\"NKGARVLVVCSEITAVTFRGPNDTHLDSLVGQALFGDGAAAVIIGSDPIPEVERPLFELV\" \\\n",
    "+\"SAAQTLLPDSEGAIDGHLREVGLTFHLLKDVPGLISKNIEKSLVEAFQPLGISDWNSLFW\" \\\n",
    "+\"IAHPGGPAILDQVELKLGLKQEKLKATRKVLSNYGNMSSACVLFILDEMRKASAKEGLGT\" \\\n",
    "+\"TGEGLEWGVLFGFGPGLTVETVVLHSVAT\", generic_protein),\n",
    "id=\"gi|13925890|gb|AAK49457.1|\",\n",
    "description=\"chalcone synthase [Nicotiana tabacum]\")\n",
    "\n",
    "my_records = [rec1, rec2, rec3]\n",
    "print(rec1)\n",
    "\n",
    "seqWritten = SeqIO.write(my_records, \"file_samples/my_proteins.fa\", \"fasta\")\n",
    "print(seqWritten, \"sequences written!\")\n",
    "print(\"\")\n",
    "print(\"Reading file\")\n",
    "seqIT = SeqIO.parse(\"file_samples/my_proteins.fa\", \"fasta\")\n",
    "\n",
    "cnt = 1\n",
    "for s in seqIT:\n",
    "    print(\"Sequence \", cnt, \"(\", s.id, \")\")\n",
    "    print(s.seq)\n",
    "    print(\"\")\n",
    "    cnt += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Multiple sequence alignment\n",
    "\n",
    "**Multiple Sequence Alignments**, as the name says, are a collection of multiple sequences which have been aligned together â usually with the insertion of gap characters, and addition of leading or trailing gaps â such that all the sequence strings are the same length. \n",
    "\n",
    "![](img/pract10/msa.gif)\n",
    "\n",
    "Such an alignment can be regarded as a matrix of letters, in biopython each row is a ```SeqRecord``` object. These alignments are stored in an object ```MultipleSeqAlignment```\n",
    "\n",
    "\n",
    "\n",
    "### Parsing multiple sequence alignments\n",
    "\n",
    "The function ```Bio.AlignIO.parse()``` returns an iterator of ```MultipleSeqAlignment``` objects. In the frequent case that we have to deal with a single alignment we will have to use the ```Bio.AlignIO.read()``` function.\n",
    "\n",
    "The basic syntax of the two functions:\n",
    "\n",
    "```Bio.AlignIO.parse(file_handle, alignment_format)```  \n",
    "```Bio.AlignIO.read(file_handle, alignment_format)```  \n",
    "\n",
    "where ```file_handle``` is the handler to the opened file, while the ```alignment_format``` is a lower case string with the alignment format (e.g. fasta, clustal, stockholm, mauve, phylip,...). Check [here](http://biopython.org/wiki/AlignIO) for all possible formats.\n",
    "\n",
    "When more than one multiple alignment has to be read we will have to use the ```parse``` function, for single entries, we would use the function ```read```.\n",
    "\n",
    "**Example:**\n",
    "Load and visualize the seed alignment of the [Piwi (PF02171) family](http://pfam.xfam.org/family/Piwi) stored in the pfam (stockholm) format. Information on the format can be found [here](https://en.wikipedia.org/wiki/Stockholm_format).  The file is available here: [PF02171_seed.sth](file_samples/PF02171_seed.sth)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SingleLetterAlphabet() alignment with 16 rows and 395 columns\n",
      "YLFFILDK-NSPEP-YGSIKRVCNTMLGVPSQCAISKHILQS--...QDV AGO1_SCHPO/500-799\n",
      "FILCILPERKTSDI-YGPWKKICLTEEGIHTQCICPIKI-----...FTK AGO6_ARATH/541-851\n",
      "FILCVLPDKKNSDL-YGPWKKKNLTEFGIVTQCMAPTRQPND--...FMK AGO4_ARATH/577-885\n",
      "CIIVVLQS-KNSDI-YMTVKEQSDIVHGIMSQCVLMKNVSRP--...HVK TAG76_CAEEL/660-966\n",
      "LIVVVLPG--KTPI-YAEVKRVGDTVLGIATQCVQAKNAIRT--...HLV O16720_CAEEL/566-867\n",
      "TFVFIITD-DSITT-LHQRYKMIEKDTKMIVQDMKLSKALSV--...LWN O62275_CAEEL/594-924\n",
      "DILVGIAR-EKKPD-VHDILKYFEESIGLQTIQLCQQTVDKMMG...NYK YQ53_CAEEL/650-977\n",
      "TIVFGIIA-EKRPD-MHDILKYFEEKLGQQTIQISSETADKFMR...NYK NRDE3_CAEEL/673-1001\n",
      "MLVVMLAD-DNKTR-YDSLKKYLCVECPIPNQCVNLRTLAGKSK...SLH Q17567_CAEEL/397-708\n",
      "IVMVVMRS-PNEEK-YSCIKKRTCVDRPVPSQVVTLKVIAPRQQ...SIN AUB_DROME/555-852\n",
      "LILCLVPN-DNAER-YSSIKKRGYVDRAVPTQVVTLKTTKNRSL...NLH PIWI_DROME/538-829\n",
      "IVVCLLSS-NRKDK-YDAIKKYLCTDCPTPSQCVVARTLGKQQT...SIH PIWL1_HUMAN/555-847\n",
      "GIMLVLPE-YNTPL-YYKLKSYLINS--IPSQFMRYDILSNRNL...VNR PIWI_ARCFU/110-406\n",
      "CFALIIGKEKYKDNDYYEILKKQLFDLKIISQNILWENWRKDDK...NWK Y1321_METJA/426-699\n",
      "LVIVFLEEYPKVDP-YKSFLLYDFVKRELLKKMIPSQVILNRTL...GIE O67434_AQUAE/419-694\n",
      "LLLAILPD-NNGSL-YGDLKRICETELGLISQCCLTKHVFKI--...YLE AGO10_ARATH/625-946\n"
     ]
    }
   ],
   "source": [
    "from Bio import AlignIO\n",
    "\n",
    "alignments = AlignIO.read(\"file_samples/PF02171_seed.sth\", \"stockholm\")\n",
    "\n",
    "print(alignments)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The code above prints a summary of the information, but each ```Bio.Align.MultipleSeqAlignment``` is composed of several alignment records (```SeqRecord```) that we can access with all their information looping through the object.\n",
    "Each ```SeqRecord``` contains several information like the **ID, Name, Description, Number of features, start, end and sequence**.\n",
    "\n",
    "**Example:**\n",
    "\n",
    "Load the seed alignment of the [Piwi (PF02171) family](http://pfam.xfam.org/family/Piwi) stored in the pfam (stockholm) format [PF02171_seed.sth](file_samples/PF02171_seed.sth). For each record print the description, start  and the end point, the sequence and the external dbxrefs (if any). "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AGO1_SCHPO/500-799 S:500 E:799\n",
      "YLFFILDK-NSPEP-YGSIKRVCNTMLGVPSQCAISKHILQS---------KPQYCANLGMKINVKVGGIN-CSLIPKSNP----LGNVPTL---------ILGGDVYHPGVGA----------TGVSIASIVASVD-LNGCKYTAVSRSQPRHQEVIEG-MKD------------IVVYLLQGFRAMTKQ-QPQRIIYFRDGTSEGQFLSVINDELSQIKEACH-------SLSPKYN--PKILVCTTQKRHHARFFIKNKSDG----------------------DRNGNPLPGTII---EKHVTHPYQYDFYLISHPSLQGVSVPVHYTVLHDEIQMPPDQF-QTL------CYNLCYVYARAT----SAVSLVPPVYYAHLVSNLARYQDV\n",
      "\n",
      "AGO6_ARATH/541-851 S:541 E:851\n",
      "FILCILPERKTSDI-YGPWKKICLTEEGIHTQCICPIKI------------SDQYLTNVLLKINSKLGGIN-SLLGIEYSYNIPLINKIPTL---------ILGMDVSHGPPGR---------ADVPSVAAVVGSKCWPLISRYRAAVRTQSPRLEMIDSLFQPIENTE--KGDNGIMNELFVEFYRTSRARKPKQIIIFRDGVSESQFEQVLKIEVDQIIKAYQ-------RLGESDV--PKFTVIVAQKNHHTKLFQAKGPE---------------------------NVPAGTVV---DTKIVHPTNYDFYMCAHAGKIGTSRPAHYHVLLDEIGFSPDDL-QNL------IHSLSYVNQRST----TATSIVAPVRYAHLAAAQVAQFTK\n",
      "\n",
      "AGO4_ARATH/577-885 S:577 E:885\n",
      "FILCVLPDKKNSDL-YGPWKKKNLTEFGIVTQCMAPTRQPND-----------QYLTNLLLKINAKLGGLN-SMLSVERTPAFTVISKVPTI---------ILGMDVSHGSPGQ---------SDVPSIAAVVSSREWPLISKYRASVRTQPSKAEMIESLVKKNGT-----EDDGIIKELLVDFYTSSNKRKPEHIIIFRDGVSESQFNQVLNIELDQIIEACK-------LLDANWN--PKFLLLVAQKNHHTKFFQPTSPE---------------------------NVPPGTII---DNKICHPKNNDFYLCAHAGMIGTTRPTHYHVLYDEIGFSADEL-QEL------VHSLSYVYQRST----SAISVVAPICYAHLAAAQLGTFMK\n",
      "\n",
      "TAG76_CAEEL/660-966 S:660 E:966\n",
      "CIIVVLQS-KNSDI-YMTVKEQSDIVHGIMSQCVLMKNVSRP---------TPATCANIVLKLNMKMGGIN--SRIVADKITNKYLVDQPTM---------VVGIDVTHPTQAEM-------RMNMPSVAAIVANVD-LLPQSYGANVKVQKKCRESVVY-LLD------------AIRERIITFYRHTKQ-KPARIIVYRDGVSEGQFSEVLREEIQSIRTACL-------AIAEDFR--PPITYIVVQKRHHARIFCKYQNDM---------------------VGKAKNVPPGTTV---DTGIVSPEGFDFYLCSHYGVQGTSRPARYHVLLDECKFTADEI-QSI------TYGMCHTYGRCT----RSVSIPTPVYYADLVATRARCHVK\n",
      "\n",
      "O16720_CAEEL/566-867 S:566 E:867\n",
      "LIVVVLPG--KTPI-YAEVKRVGDTVLGIATQCVQAKNAIRT---------TPQTLSNLCLKMNVKLGGVN-SILLPNVRPR---IFNEPVI---------FLGCDITHPAAGD---------TRKPSIAAVVGSMD-AHPSRYAATVRVQQHRQEIITD-LTY------------MVRELLVQFYRNTRF-KPARIVVYRDGVSEGQLFNVLQYELRAIREACV-------MLESGYQ--PGITFIAVQKRHHTRLFAADKADQ---------------------VGKAFNIPPGTTV---DVGITHPTEFDFFLCSHAGIQGTSRPSHYHVLWDDNDLTADEL-QQL------TYQMCHTYVRCT----RSVSIPAPAYYAHLVAFRARYHLV\n",
      "\n",
      "O62275_CAEEL/594-924 S:594 E:924\n",
      "TFVFIITD-DSITT-LHQRYKMIEKDTKMIVQDMKLSKALSV--IN---AGKRLTLENVINKTNVKLGGSN--YVFVDAKKQL-----DSHL---------IIGVGISAPPAGTKYAM-ENKGVLNPNVIGYAYNA--QHNQEFSGDFVLNSASQDTLAP-IED------------IVMHSLNEYQKFHDGGLPRRVIVYRTGTSEGNHGSIMAYEIPLARAAMR-------DFSPDIQ----LVYIVVSKDHSFRFFKPDLASLASRPQATSSTASRHSAMPAAPKAWDLNIAPGILV---DSIVTNPACKQFFLNSHITLQGTAKTPLYTVLADDAKVSMTAL-EDI------TYKLCHLHQIVG----LPTSLPTPLYVANEYAKRGRNLWN\n",
      "\n",
      "YQ53_CAEEL/650-977 S:650 E:977\n",
      "DILVGIAR-EKKPD-VHDILKYFEESIGLQTIQLCQQTVDKMMGG----QGGRQTIDNVMRKFNLKCGGTNFFVEIPNAVRGKAVCSNNETLRKKLLEHVQFIGFEISHGASRTLFDRSRSQMDGEPSVVGVSYSLT---NSTQLGGFTYLQTQKEYKLQKLDE------------FFPKCVRSYKEHSKT-LPTRIVIYRVGAGEGNFNRVKE-EVEEMRRTFD-------KIQPGYR--PHLVVIIAQRASHARVFPSCISGN---------------------RATDQNIPSGTCV---ENVLTSYGYDEFILSSQTPLIGTVRPCKYTILVNDAKWSKNEL-MHL------TYFRAFGHQVSY----QPPSVPDVLYAAENLAKRGRNNYK\n",
      "\n",
      "NRDE3_CAEEL/673-1001 S:673 E:1001\n",
      "TIVFGIIA-EKRPD-MHDILKYFEEKLGQQTIQISSETADKFMRD----HGGKQTIDNVIRKLNPKCGGTNFLIDVPESVGHRVVCNNSAEMRAKLYAKTQFIGFEMSHTGARTRFDIQKVMFDGDPTVVGVAYSL--KHSAQLGGFSYFQESRLHKLTN-LQE------------KMQICLNAYEQSSSY-LPETVVVYRVGSGEGDYPQIVN-EVNEMKLAAR-------KKKHGYN--PKFLVICTQRNSHIRVFPEHINERG--------------------KSMEQNVKSGTCV---DVPGASHGYEEFILCCQTPLIGTVKPTKYTIIVNDCRWSKNEI-MNV------TYHLAFAHQVSY----APPAIPNVSYAAQNLAKRGHNNYK\n",
      "\n",
      "Q17567_CAEEL/397-708 S:397 E:708\n",
      "MLVVMLAD-DNKTR-YDSLKKYLCVECPIPNQCVNLRTLAGKSKDGGENKNLGSIVLKIVLQMICKTGGA---LWKVNIPL-------KSTM---------IVGYDLYHDSTLK-----------GKTVGACVSTTS-NDFTQFYSQTRPHENPTQLGNN-LTH------------FVRKSLKQYYDNNDKTLPSRLILYRDGAGDGQIPYIKNTEVKLVRDACDAVTDKAAELSNKVQEKIKLAFIIVTKRVNMRILKQGSSS-----------------------KSAINPQPGTVV---DTTVTRPERMDFYLVPQFVNQGTVTPVSYNIIHDDTGLGPDKH-QQL------AFKLCHLYYNWQ----GTVRVPAPCQYAHKLAFLTAQSLH\n",
      "\n",
      "AUB_DROME/555-852 S:555 E:852\n",
      "IVMVVMRS-PNEEK-YSCIKKRTCVDRPVPSQVVTLKVIAPRQQKP---TGLMSIATKVVIQMNAKLMGA---PWQVVIPL-------HGLM---------TVGFDVCHSPKNK-----------NKSYGAFVATMDQKESFRYFSTVNEHIKGQELSEQ-MSV------------NMACALRSYQEQHRS-LPERILFFRDGVGDGQLYQVVNSEVNTLKDRLDEI---YKSAGKQEG--CRMTFIIVSKRINSRYFT-----------------------------GHRNPVPGTVV---DDVITLPERYDFFLVSQAVRIGTVSPTSYNVISDNMGLNADKL-QML------SYKMTHMYYNYS----GTIRVPAVCHYAHKLAFLVAESIN\n",
      "\n",
      "PIWI_DROME/538-829 S:538 E:829\n",
      "LILCLVPN-DNAER-YSSIKKRGYVDRAVPTQVVTLKTTKNRSL--------MSIATKIAIQLNCKLGYT---PWMIELPL-------SGLM---------TIGFDIAKSTRDR-----------KRAYGALIASMDLQQNSTYFSTV-TECSAFDVLANTLWP------------MIAKALRQYQHEHRK-LPSRIVFYRDGVSSGSLKQLFEFEVKDIIEKLKTE---YARVQLSP---PQLAYIVVTRSMNTRFFLN-----------------------------GQNPPPGTIV---DDVITLPERYDFYLVSQQVRQGTVSPTSYNVLYSSMGLSPEKM-QKL------TYKMCHLYYNWS----GTTRVPAVCQYAKKLATLVGTNLH\n",
      "\n",
      "PIWL1_HUMAN/555-847 S:555 E:847\n",
      "IVVCLLSS-NRKDK-YDAIKKYLCTDCPTPSQCVVARTLGKQQT-------VMAIATKIALQMNCKMGGE---LWRVDIPL-------KLVM---------IVGIDCYHDMTAG-----------RRSIAGFVASIN-EGMTRWFSRCIFQDRGQELVDG-LKV------------CLQAALRAW-NSCNEYMPSRIIVYRDGVGDGQLKTLVNYEVPQFLDCLK-------SIGRGYN--PRLTVIVVKKRVNTRFFAQSGGR-------------------------LQNPLPGTVI---DVEVTRPEWYDFFIVSQAVRSGSVSPTHYNVIYDNSGLKPDHI-QRL------TYKLCHIYYNWP----GVIRVPAPCQYAHKLAFLVGQSIH\n",
      "\n",
      "PIWI_ARCFU/110-406 S:110 E:406\n",
      "PDB; 2W42 B; 110-406;,PDB; 1YTU B; 110-406;,PDB; 2BGG B; 110-406;,PDB; 1W9H A; 110-406;,PDB; 2BGG A; 110-406;,PDB; 1YTU A; 110-406;,PDB; 2W42 A; 110-406;\n",
      "GIMLVLPE-YNTPL-YYKLKSYLINS--IPSQFMRYDILSNRNL--------TFYVDNLLVQFVSKLGGK---PWILNVDPEK--------------------GSDIIIGTGAT--------RIDNVNLFCFAMVFK-KDGTMLWNEISPIVTSSEYLTY-LKS------------TIKKVVYGFKKSNPDWDVEKLTLHVSG----KRPKMKDGETKILKETVE-------ELKKQEMVSRDVKYAILHLNETHPFWVMGDPN------------------------NRFHPYEGTKVKLSSKRYLLTLLQPYLKRNGLEMVTPIKPLSVEIVSDN--WTSEEYYHNVHEILDEIYYLSKMNWRGF----RSRNLPVTVNYPKLVAGIIANVNR\n",
      "\n",
      "Y1321_METJA/426-699 S:426 E:699\n",
      "CFALIIGKEKYKDNDYYEILKKQLFDLKIISQNILWENWRKDDK--------GYMTNNLLIQIMGKLGIK---YFILDSKTPYDY----------------IMGLDTGLGIFGN------------HRVGGCTVVYDSEGKIRRIQPIETPAPGERLHLP----------------YVIEYLEN--KANIDMENKNILFLRDG-------FIQNSERNDLKEISK-------ELN------SNIEVISIRKNNKYKVFTSDYRI------------------------GSVFGNDGIFLPHKTPFGSNPVKLSTWLRFNCGNEEGLK---IN----------ESI-MQL------LYDLTKMNYSALYGEGRYLRIPAPIHYADKFVKALGKNWK\n",
      "\n",
      "O67434_AQUAE/419-694 S:419 E:694\n",
      "PDB; 1YVU A; 419-694;,PDB; 2F8S A; 419-694;,PDB; 2F8T A; 419-694;,PDB; 2F8S B; 419-694;,PDB; 2NUB A; 419-694;,PDB; 2F8T B; 419-694;\n",
      "LVIVFLEEYPKVDP-YKSFLLYDFVKRELLKKMIPSQVILNRTLKN---ENLKFVLLNVAEQVLAKTGNIP--YKLKEIEGKVDA----------------FVGIDISRITRDG----------KTVNAVAFTKIFNSKGELVRYYLTSYPAFGEKLTEK----------------AIGDVFSLLEKLGF-KKGSKIVVHRDG-------RLYRDEVAAFK-----------KYGELYG--YSLELLEIIKRNNPRFFSNEKFI------------------------------KGYFYKLSEDSVILATYNQVY-------EGTHQPIKVRKVYGE--LPVEVL-CSQ------ILSLTLMNYSSF----QPIKLPATVHYSDKITKLMLRGIE\n",
      "\n",
      "AGO10_ARATH/625-946 S:625 E:946\n",
      "LLLAILPD-NNGSL-YGDLKRICETELGLISQCCLTKHVFKI---------SKQYLANVSLKINVKMGGRN-TVLVDAISCRIPLVSDIPTI---------IFGADVTHPENGE---------ESSPSIAAVVASQDWPEVTKYAGLVCAQAHRQELIQDLYKTWQDPVRGTVSGGMIRDLLISFRKATGQ-KPLRIIFYRDGVSEGQFYQVLLYELDAIRKACA-------SLEPNYQ--PPVTFIVVQKRHHTRLFANNHRDKN-------------------STDRSGNILPGTVV---DTKICHPTEFDFYLCSHAGIQGTSRPAHYHVLWDENNFTADGI-QSL------TNNLCYTYARCT----RSVSIVPPAYYAHLAAFRARFYLE\n",
      "\n"
     ]
    }
   ],
   "source": [
    "from Bio import AlignIO\n",
    "\n",
    "alignments = AlignIO.read(\"file_samples/PF02171_seed.sth\", \"stockholm\")\n",
    "\n",
    "\n",
    "for align in alignments:\n",
    "    start = align.annotations[\"start\"]\n",
    "    end = align.annotations[\"end\"]\n",
    "    seq = align.seq\n",
    "    desc = align.description\n",
    "    dbref = \",\".join([x for x in align.dbxrefs])\n",
    "    print(\"{} S:{} E:{}\".format(desc, start, end))\n",
    "    if(len(dbref) > 0):\n",
    "        print(dbref)\n",
    "    print(\"{}\".format(seq))\n",
    "    print(\"\")\n",
    "                \n",
    "    \n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Writing multiple alignments\n",
    "\n",
    "Biopython provides a function ```Bio.AlignIO.write()``` to write alignments to file. The basic syntax is:\n",
    "```\n",
    "N = Bio.AlignIO.write(alignments,outfile,file_format)\n",
    "```\n",
    "where ```alignments``` are a MultipleSeqAlignment object with the alignments to write to the output file with name ```outfile``` that has format ```file_format``` (a low case string with the file format). N is the number of entries written to the file.\n",
    "\n",
    "**Example:**\n",
    "\n",
    "Let's create some multiple sequence alignments and store them in [phylip](http://www.phylo.org/tools/obsolete/phylip.html) format. The output file is present here [my_malign.phy](file_samples/my_malign.phy). Finally, read and print the content of the file.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3 entries written to the file\n",
      "\n",
      "SingleLetterAlphabet() alignment with 3 rows and 12 columns\n",
      "ACTGCTAGCTAG Alpha\n",
      "ACT-CTAGCTAG Beta\n",
      "ACTGCTAGDTAG Gamma\n",
      "SingleLetterAlphabet() alignment with 3 rows and 9 columns\n",
      "GTCAGC-AG Delta\n",
      "GACAGCTAG Epsilon\n",
      "GTCAGCTAG Zeta\n",
      "SingleLetterAlphabet() alignment with 3 rows and 13 columns\n",
      "ACTAGTACAGCTG Eta\n",
      "ACTAGTACAGCT- Theta\n",
      "-CTACTACAGGTG Iota\n"
     ]
    }
   ],
   "source": [
    "\n",
    "from Bio.Alphabet import generic_dna\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.Align import MultipleSeqAlignment\n",
    "from Bio import AlignIO\n",
    "\n",
    "align1 = MultipleSeqAlignment([\n",
    "SeqRecord(Seq(\"ACTGCTAGCTAG\", generic_dna), id=\"Alpha\"),\n",
    "SeqRecord(Seq(\"ACT-CTAGCTAG\", generic_dna), id=\"Beta\"),\n",
    "SeqRecord(Seq(\"ACTGCTAGDTAG\", generic_dna), id=\"Gamma\"),\n",
    "])\n",
    "align2 = MultipleSeqAlignment([\n",
    "SeqRecord(Seq(\"GTCAGC-AG\", generic_dna), id=\"Delta\"),\n",
    "SeqRecord(Seq(\"GACAGCTAG\", generic_dna), id=\"Epsilon\"),\n",
    "SeqRecord(Seq(\"GTCAGCTAG\", generic_dna), id=\"Zeta\"),\n",
    "])\n",
    "align3 = MultipleSeqAlignment([\n",
    "SeqRecord(Seq(\"ACTAGTACAGCTG\", generic_dna), id=\"Eta\"),\n",
    "SeqRecord(Seq(\"ACTAGTACAGCT-\", generic_dna), id=\"Theta\"),\n",
    "SeqRecord(Seq(\"-CTACTACAGGTG\", generic_dna), id=\"Iota\"),\n",
    "])\n",
    "\n",
    "my_alignments = [align1, align2, align3]\n",
    "N = AlignIO.write(my_alignments, \"file_samples/my_malign.phy\", \"phylip\")\n",
    "\n",
    "print(\"{} entries written to the file\".format(N))\n",
    "print(\"\")\n",
    "alignments = AlignIO.parse(\"file_samples/my_malign.phy\", \"phylip\")\n",
    "for align in alignments:\n",
    "    print(align)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Note that it is also possible to convert one format into the other (provided that all information needed for the second format is available) by using the function:\n",
    "```\n",
    "Bio.AlignIO.convert(input_file, input_file_format, output_file, output_file_format)\n",
    "```\n",
    "basically by passing the input file name and format and output file name and format.\n",
    "\n",
    "Ex:\n",
    "```\n",
    "Bio.AlignIO.convert(\"PF05371_seed.sth\", \"stockholm\", \"PF05371_seed.aln\", \"clustal\")\n",
    "```\n",
    "converts a pfam file into a clustal. \n",
    "\n",
    "\n",
    "### Manipulating alignments\n",
    "\n",
    "It is possible to slice alignments using the [] operator applied on a ```SeqRecord```. In particular:\n",
    "\n",
    "1. ```SeqRecord[i,j]``` returns the jth character of alignment i as a string;\n",
    "2. ```SeqRecord[:,j]``` returns all the jth characters of the multiple alignment as a string;\n",
    "3. ```SeqRecord[:,i:j]``` returns a MultipleSeqAlignment with the sub-alignments going for i to j (excluded)\n",
    "4. ```SeqRecord[a:b,i:j]``` similar to 3. but for alignments going from a to b (excluded) only\n",
    "\n",
    "One can also combine different portions of the alignment by slicing and adding the slices together with\n",
    "```\n",
    "SeqRecord[:,i:j] + SeqRecord[:,t:k]\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Example:**\n",
    "\n",
    "Load the seed alignment of the [Piwi (PF02171) family](http://pfam.xfam.org/family/Piwi) stored in the pfam (stockholm) format [PF02171_seed.sth](file_samples/PF02171_seed.sth). For each record print the description, the first 20 bases and the final 20 bases of the sequence. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AGO1_SCHPO/500-799 \t YLFFILDK-NSPEP-YGSIKLVPPVYYAHLVSNLARYQDV \t\n",
      "AGO6_ARATH/541-851 \t FILCILPERKTSDI-YGPWKIVAPVRYAHLAAAQVAQFTK \t\n",
      "AGO4_ARATH/577-885 \t FILCVLPDKKNSDL-YGPWKVVAPICYAHLAAAQLGTFMK \t\n",
      "TAG76_CAEEL/660-966 \t CIIVVLQS-KNSDI-YMTVKIPTPVYYADLVATRARCHVK \t\n",
      "O16720_CAEEL/566-867 \t LIVVVLPG--KTPI-YAEVKIPAPAYYAHLVAFRARYHLV \t\n",
      "O62275_CAEEL/594-924 \t TFVFIITD-DSITT-LHQRYLPTPLYVANEYAKRGRNLWN \t\n",
      "YQ53_CAEEL/650-977 \t DILVGIAR-EKKPD-VHDILVPDVLYAAENLAKRGRNNYK \t\n",
      "NRDE3_CAEEL/673-1001 \t TIVFGIIA-EKRPD-MHDILIPNVSYAAQNLAKRGHNNYK \t\n",
      "Q17567_CAEEL/397-708 \t MLVVMLAD-DNKTR-YDSLKVPAPCQYAHKLAFLTAQSLH \t\n",
      "AUB_DROME/555-852 \t IVMVVMRS-PNEEK-YSCIKVPAVCHYAHKLAFLVAESIN \t\n",
      "PIWI_DROME/538-829 \t LILCLVPN-DNAER-YSSIKVPAVCQYAKKLATLVGTNLH \t\n",
      "PIWL1_HUMAN/555-847 \t IVVCLLSS-NRKDK-YDAIKVPAPCQYAHKLAFLVGQSIH \t\n",
      "PIWI_ARCFU/110-406 \t GIMLVLPE-YNTPL-YYKLKLPVTVNYPKLVAGIIANVNR \t\n",
      "Y1321_METJA/426-699 \t CFALIIGKEKYKDNDYYEILIPAPIHYADKFVKALGKNWK \t\n",
      "O67434_AQUAE/419-694 \t LVIVFLEEYPKVDP-YKSFLLPATVHYSDKITKLMLRGIE \t\n",
      "AGO10_ARATH/625-946 \t LLLAILPD-NNGSL-YGDLKIVPPAYYAHLAAFRARFYLE \t\n"
     ]
    }
   ],
   "source": [
    "from Bio import AlignIO\n",
    "\n",
    "alignments = AlignIO.read(\"file_samples/PF02171_seed.sth\", \"stockholm\")\n",
    "\n",
    "edited_aligns = alignments[:,0:20] + alignments[:,-20:]\n",
    "\n",
    "\n",
    "for align in edited_aligns:\n",
    "    print(align.description, \"\\t\", align.seq, \"\\t\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Biopython's pairwise2 alignment\n",
    "\n",
    "Biopython supports running some classic multiple alignment tools like [clustal](http://www.sciencedirect.com/science/article/pii/0378111988903307?via%3Dihub), [MUSCLE](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC390337/) but these require to be installed in your system first. \n",
    "\n",
    "Biopython has its own module to make pairwise alignment. It provides two algorithms: [Smith-Waterman](https://en.wikipedia.org/wiki/Smith%E2%80%93Waterman_algorithm) for local alignment and [Needleman-Wunsch](https://en.wikipedia.org/wiki/Needleman%E2%80%93Wunsch_algorithm) for global alignment. These methods are implemented in two Biopython functions of the ```Bio.pairwise2``` module: \n",
    "\n",
    "```pairwise2.align.globalxx()``` and ```pairwise2.align.localxx()```.\n",
    "\n",
    "The basic syntax of the two methods is the same:\n",
    "```\n",
    "aligns = pairwise2.align.globalxx(seq1,seq2)\n",
    "aligns = pairwise2.align.localxx(seq1,seq2)\n",
    "```\n",
    "where seq1 and seq2 are two ```str``` objects. These methods return a list of alignments (at least one) that have the same **optimal score**. Each alignment is represented as tuples with the following 5 elements in order:\n",
    "\n",
    "1. The alignment of the first sequence;\n",
    "2. The alignment of the second sequence;\n",
    "3. The alignment score;\n",
    "4. The start of the alignment (for global alignments this is always 0);\n",
    "5. The end of the alignment (for global alignments this is always the length of the alignment).\n",
    "\n",
    "The **xx** in ```globalxx``` or ```localxx``` encodes the penalties and scores for gaps and matches (and mismatches) of the alignment. The first letter is **the score for a match** the second letter is **the penalty for a gap**:\n",
    "\n",
    "Match parameters can be:\n",
    "- ```x``` : means that a match scores 1 a mismatch 0;\n",
    "- ```m``` : the match and mismatch score are passed as additional params after the sequence (es. ```aligns = pairwise2.align.globalmx(seq1,seq2, 1, -1)``` to set 1 as match score and -1 as mismatch penalty.\n",
    "\n",
    "Gap parameters can be:\n",
    "- ```x``` : gap penalty is 0;\n",
    "- ```s``` : same gap open and gap extend penalties. They are passed as additional params after the sequence.\n",
    "- ```d``` : different gap open and gap extend penalties. They are passed as additional params after the sequence.\n",
    "\n",
    "**Example:**\n",
    "\n",
    "Let's perform the alignment of the two sequences \"ACCGTTATATAGGCCA\" and \"ACGTACTAGTATAGGCCA\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[('ACCGT--TA-TATAGGCCA', 'A-CGTACTAGTATAGGCCA', 15.0, 0, 19), ('ACCGT--TA-TATAGGCCA', 'AC-GTACTAGTATAGGCCA', 15.0, 0, 19)]\n",
      "\n",
      "Looping through aligns\n",
      "ACCGT--TA-TATAGGCCA\n",
      "A-CGTACTAGTATAGGCCA\n",
      "Score: 15.0, Start: 0, End: 19\n",
      "\n",
      "ACCGT--TA-TATAGGCCA\n",
      "AC-GTACTAGTATAGGCCA\n",
      "Score: 15.0, Start: 0, End: 19\n",
      "\n",
      "\n",
      "Match: 1, Mismatch: -1, Gap open: -0.5, Gap extend: -0.2\n",
      "ACCGT--TA-TATAGGCCA\n",
      "A-CGTACTAGTATAGGCCA\n",
      "Score: 13.3, Start: 0, End: 19\n",
      "\n",
      "ACCGT--TA-TATAGGCCA\n",
      "AC-GTACTAGTATAGGCCA\n",
      "Score: 13.3, Start: 0, End: 19\n",
      "\n"
     ]
    }
   ],
   "source": [
    "from Bio import pairwise2\n",
    "from Bio import SeqIO\n",
    "from Bio.pairwise2 import format_alignment\n",
    "\n",
    "\n",
    "alignments = pairwise2.align.globalxx(\"ACCGTTATATAGGCCA\", \"ACGTACTAGTATAGGCCA\")\n",
    "print(alignments)\n",
    "\n",
    "print(\"\")\n",
    "print(\"Looping through aligns\")\n",
    "for align in alignments:\n",
    "        print(align[0])\n",
    "        print(align[1])\n",
    "        print(\"Score: {}, Start: {}, End: {}\".format(align[2],align[3],align[4]))\n",
    "        print(\"\")\n",
    "\n",
    "alignments = pairwise2.align.globalms(\"ACCGTTATATAGGCCA\", \"ACGTACTAGTATAGGCCA\",1,-1,-0.5,-0.2)\n",
    "print(\"\")\n",
    "print(\"Match: 1, Mismatch: -1, Gap open: -0.5, Gap extend: -0.2\")\n",
    "for align in alignments:\n",
    "    print(align[0])\n",
    "    print(align[1])\n",
    "    print(\"Score: {}, Start: {}, End: {}\".format(align[2],align[3],align[4]))\n",
    "    print(\"\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "It is also possible to specify a substitution matrix like the [BLOSUM](https://en.wikipedia.org/wiki/BLOSUM) to improve the alignment of protein sequences. Substitution matrices like the **blosum62** is present in Biopython and before using it we need to import it with ``` from Bio.SubsMat.MatrixInfo import blosum62```.\n",
    "\n",
    "You can find all the substitution matrices available [here](http://biopython.org/DIST/docs/api/). \n",
    "\n",
    "**Example:**\n",
    "Align the alpha and beta subunits of the human hemoglobin protein stored in [hem_alpha.fasta](file_samples/hem_alpha.fasta) and [hem_beta.fasta](file_samples/hem_beta.fasta) with globalxx and with a blosum62 substitution matrix and gap open -10 and gap extension -0.5. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "80 optimal alignments\n",
      "MV-LSPADKTNV---K-A--A-WGKVGAHAG---EY-GA-EALE-RMFLSF----PTTK-TY--FPHFDL-SH-G--S---AQVK-G------HGKKV--A--DA-LTNAVAHVD-DMPNALS----A-LSD-LHAH--KLR-VDPV-NFKL-LSHCL---LVT--LAAHLPA----EFTPA-VH-ASLDKFLAS---VSTV------LTS--KYR-\n",
      "MVHL-----T--PEEKSAVTALWGKV-----NVDE-VG-GEAL-GR--L--LVVYP---WT-QRF--F--ES-FGDLSTPDA-V-MGNPKVKAHGKKVLGAFSD-GL----AH--LD--N-L-KGTFATLS-EL--HCDKL-HVDP-ENF--RL---LGNVLV-CVL-AH---HFGKEFTP-PV-QA------A-YQKV--VAGVANAL--AHKY-H\n",
      "Score: 72.0, Start: 0, End: 217\n",
      "\n",
      "2 optimal alignments\n",
      "MV-LSPADKTNVKAAWGKVGAHAGEYGAEALERMFLSFPTTKTYFPHF-DLS-----HGSAQVKGHGKKVADALTNAVAHVDDMPNALSALSDLHAHKLRVDPVNFKLLSHCLLVTLAAHLPAEFTPAVHASLDKFLASVSTVLTSKYR\n",
      "MVHLTPEEKSAVTALWGKV--NVDEVGGEALGRLLVVYPWTQRFFESFGDLSTPDAVMGNPKVKAHGKKVLGAFSDGLAHLDNLKGTFATLSELHCDKLHVDPENFRLLGNVLVCVLAHHFGKEFTPPVQAAYQKVVAGVANALAHKYH\n",
      "Score: 292.5, Start: 0, End: 149\n"
     ]
    }
   ],
   "source": [
    "from Bio import pairwise2\n",
    "from Bio import SeqIO\n",
    "from Bio.pairwise2 import format_alignment\n",
    "from Bio.SubsMat.MatrixInfo import blosum62\n",
    "\n",
    "\n",
    "\n",
    "seq1 = SeqIO.read(\"file_samples/hem_alpha.fasta\", \"fasta\")\n",
    "seq2 = SeqIO.read(\"file_samples/hem_beta.fasta\", \"fasta\")\n",
    "\n",
    "\n",
    "alignments = pairwise2.align.globalxx(seq1.seq,seq2.seq)\n",
    "print(len(alignments), \"optimal alignments\")\n",
    "\n",
    "align = alignments[0]\n",
    "print(align[0])\n",
    "print(align[1])\n",
    "print(\"Score: {}, Start: {}, End: {}\".format(align[2],align[3],align[4]))\n",
    "print(\"\")\n",
    "\n",
    "alignments = pairwise2.align.globalds(seq1.seq, seq2.seq, blosum62, -10, -0.5)\n",
    "print(len(alignments), \"optimal alignments\")\n",
    "\n",
    "align = alignments[0]\n",
    "print(align[0])\n",
    "print(align[1])\n",
    "print(\"Score: {}, Start: {}, End: {}\".format(align[2],align[3],align[4]))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Exercises\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "1. Write a python function that reads a genebank file given in input and prints off the following information:\n",
    "    1. Identifier, name and description;\n",
    "    2. The first 100 characters of the sequence;\n",
    "    3. Number of external references (dbxrefs) and ids of the external refs.\n",
    "    4. The name of the organism (hint: check the annotations dictionary at the key \"organism\")\n",
    "    5. Retrieve and print all (if any) associated publications (hint: annotation dictionary, key:\"references\")\n",
    "    6. Retrieve and print all the locations of \"CDS\" features of the sequence (hint: check the features )\n",
    "   \n",
    "Hint: go back and check the details of the ```SeqRecord``` object. \n",
    "\n",
    "Test the program downloading some files from genebank like [this](https://www.ncbi.nlm.nih.gov/nuccore/AB026718.1)\n",
    "    \n",
    "<div class=\"tggle\" onclick=\"toggleVisibility('ex1');\">Show/Hide Solution</div>\n",
    "<div id=\"ex1\" style=\"display:none;\">"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ID:NC_005816.1\n",
      "NAME:NC_005816\n",
      "DESC:Yersinia pestis biovar Microtus str. 91001 plasmid pPCP1, complete sequence\n",
      "ORGANISM:Yersinia pestis biovar Microtus str. 91001\n",
      "\n",
      "CDS features:\n",
      "CDS [86:1109](+)\n",
      "CDS [1105:1888](+)\n",
      "CDS [2924:3119](+)\n",
      "CDS [3485:3857](+)\n",
      "CDS [4342:4780](+)\n",
      "CDS [4814:5888](-)\n",
      "CDS [6004:6421](+)\n",
      "CDS [6663:7602](+)\n",
      "CDS [7788:8088](-)\n",
      "CDS [8087:8360](-)\n",
      "num DBrefs 1:\n",
      "Refs:Project:58037\n",
      "\n",
      "PUBLICATIONS:\n",
      "REFERENCE [1]\n",
      "Genetics of metabolic variations between Yersinia pestis biovars and the proposal of a new biovar, microtus\n",
      "Zhou,D., Tong,Z., Song,Y., Han,Y., Pei,D., Pang,X., Zhai,J., Li,M., Cui,B., Qi,Z., Jin,L., Dai,R., Du,Z., Wang,J., Guo,Z., Wang,J., Huang,P. and Yang,R.\n",
      "J. Bacteriol. 186 (15), 5147-5152 (2004)\n",
      "\n",
      "REFERENCE [2]\n",
      "Complete genome sequence of Yersinia pestis strain 91001, an isolate avirulent to humans\n",
      "Song,Y., Tong,Z., Wang,J., Wang,L., Guo,Z., Han,Y., Zhang,J., Pei,D., Zhou,D., Qin,H., Pang,X., Han,Y., Zhai,J., Li,M., Cui,B., Qi,Z., Jin,L., Dai,R., Chen,F., Li,S., Ye,C., Du,Z., Lin,W., Wang,J., Yu,J., Yang,H., Wang,J., Huang,P. and Yang,R.\n",
      "DNA Res. 11 (3), 179-197 (2004)\n",
      "\n",
      "REFERENCE [3]\n",
      "Direct Submission\n",
      "\n",
      "Submitted (16-MAR-2004) National Center for Biotechnology Information, NIH, Bethesda, MD 20894, USA\n",
      "\n",
      "REFERENCE [4]\n",
      "Direct Submission\n",
      "Song,Y., Tong,Z., Wang,L., Han,Y., Zhang,J., Pei,D., Wang,J., Zhou,D., Han,Y., Pang,X., Zhai,J., Chen,F., Qin,H., Wang,J., Li,S., Guo,Z., Ye,C., Du,Z., Lin,W., Wang,J., Yu,J., Yang,H., Wang,J., Huang,P. and Yang,R.\n",
      "Submitted (24-APR-2003) The Institute of Microbiology and Epidemiology, Academy of Military Medical Sciences, No. 20, Dongdajie Street, Fengtai District, Beijing 100071, People's Republic of China\n",
      "\n"
     ]
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "\n",
    "def getGenBankInfo(filename): \n",
    "    record = SeqIO.read(filename, \"genbank\")\n",
    "    \n",
    "    ident = record.id\n",
    "    name = record.name\n",
    "    desc = record.description\n",
    "    numDbrefs = len(record.dbxrefs)\n",
    "    dbrefs = \",\".join(record.dbxrefs)\n",
    "    organism = record.annotations[\"organism\"]\n",
    "    pubs = record.annotations[\"references\"]\n",
    "    \n",
    "    print(\"ID:{}\".format(ident) )\n",
    "    print(\"NAME:{}\".format(name))\n",
    "    print(\"DESC:{}\".format(desc) )\n",
    "    print(\"ORGANISM:{}\".format(organism))\n",
    "    print(\"\")\n",
    "    print(\"CDS features:\")\n",
    "    for f in record.features:\n",
    "        if(f.type == \"CDS\"):\n",
    "            print(f.type, f.location)\n",
    "    print(\"num DBrefs {}:\".format(numDbrefs))\n",
    "    if(numDbrefs > 0):\n",
    "        print(\"Refs:{}\".format(dbrefs))\n",
    "    print(\"\")\n",
    "    if(len(pubs) > 0):\n",
    "        print(\"PUBLICATIONS:\")\n",
    "        pcount = 1\n",
    "        for p in pubs:\n",
    "            print(\"REFERENCE [{}]\".format(pcount))\n",
    "            print(\"{}\".format(p.title))\n",
    "            print(\"{}\".format(p.authors))\n",
    "            print(\"{}\".format(p.journal))\n",
    "            print(\"\")\n",
    "            pcount += 1\n",
    "\n",
    "\n",
    "        \n",
    "fn = \"file_samples/NC_005816.gb\"\n",
    "getGenBankInfo(fn)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "2. Write a python program that loads a pfam file (stockholm format .sth) and reports for each record of the alignment:\n",
    "    1. the id of the entry\n",
    "    2. the start and end points\n",
    "    3. the number of gaps and the % of gaps on the total length of the alignment\n",
    "    4. the number of external database references (dbxrefs), and the first 3 external references comma separated (hint: use join).\n",
    "\n",
    "Print these information to the screen. Finally, write this information in a tab separated file (.tsv) having the following format:\n",
    "```#ID\\tstart\\tend\\tnum_gaps\\tpercentage_gaps\\tdbxrefs```.\n",
    "\n",
    "You can test the program using the [TP53](https://en.wikipedia.org/wiki/TP53) family file [PF00870_seed.sth](file_samples/PF00870_seed.sth).\n",
    "<div class=\"tggle\" onclick=\"toggleVisibility('ex2');\">Show/Hide Solution</div>\n",
    "<div id=\"ex2\" style=\"display:none;\">"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SingleLetterAlphabet() alignment with 5 rows and 197 columns\n",
      "SSVPSQKTYQGSYGFRLGFLHSGTAKSVTCTYSPALNKMFCQLA...ENL P53_HUMAN/95-289\n",
      "STVPETSDYPGDHGFRLRFPQSGTAKSVTCTYSPDLNKLFCQLA...SNF P53_DANRE/63-257\n",
      "-AVPSTEDYAGSYGLKLEFQQNGTAKSVTCTYSTDLNKLFCQLA...DN- F7A9U0_XENTR/70-263\n",
      "PVIPSNTDYPGPHHFEVTFQQSSTAKSATWTYSPLLKKLYCQIA...DHY P73_HUMAN/113-309\n",
      "TTVPVTTDYPGSYELELRFQKSGTAKSVTSTYSETLNKLYCQLA...ESR P53_ORYLA/80-270\n",
      "ID: P53_HUMAN/95-289 start: 95 end: 289 gaps: 2 % gaps: 1.03 refs: 210 PDB; 3Q06 B; 96-289;,PDB; 4MZR B; 95-289;,PDB; 5LAP B; 95-289;\n",
      "ID: P53_DANRE/63-257 start: 63 end: 257 gaps: 2 % gaps: 1.03 refs: 0 \n",
      "ID: F7A9U0_XENTR/70-263 start: 70 end: 263 gaps: 3 % gaps: 1.55 refs: 0 \n",
      "ID: P73_HUMAN/113-309 start: 113 end: 309 gaps: 0 % gaps: 0.00 refs: 45 PDB; 4A63 E; 114-309;,PDB; 4G83 A; 115-309;,PDB; 3VD2 A; 115-309;\n",
      "ID: P53_ORYLA/80-270 start: 80 end: 270 gaps: 6 % gaps: 3.16 refs: 0 \n",
      "\n",
      "Written 5 entries.\n"
     ]
    }
   ],
   "source": [
    "from Bio import AlignIO\n",
    "\n",
    "def getStockholmInfo(fn):\n",
    "\n",
    "    alignments = AlignIO.read(fn, \"stockholm\")\n",
    "    print(alignments)\n",
    "    retVals = []\n",
    "    for align in alignments:\n",
    "        ident = align.id\n",
    "        start = align.annotations[\"start\"]\n",
    "        end = align.annotations[\"end\"]\n",
    "        gaps = align.seq.count(\"-\")\n",
    "        #percentage gaps\n",
    "        percGaps = 100*gaps/(end - start)\n",
    "        dbxrfsN = len(align.dbxrefs)\n",
    "        #get top five external refs\n",
    "        dbxrfs = \",\".join(align.dbxrefs[0:3])\n",
    "        print(\"ID: {} start: {} end: {} gaps: {} % gaps: {:.2f} refs: {} {}\".format(\n",
    "            ident, start, end, gaps, percGaps, dbxrfsN, dbxrfs))\n",
    "        retVals.append([ident, start, end, gaps, \"{:.2f}\".format(percGaps), dbxrfsN, dbxrfs])\n",
    "    return retVals\n",
    "\n",
    "def writeStockholmInfo(outFile, data):\n",
    "    fh = open(outFile, \"w\")\n",
    "    fh.write(\"#ID\\tstart\\tend\\tnum_gaps\\tpercentage_gaps\\tdbxrefs\\n\")\n",
    "    for entry in data:\n",
    "        fh.write(\"\\t\".join([str(x) for x in entry]) + \"\\n\")\n",
    "    fh.close()\n",
    "    print(\"Written {} entries.\".format(len(data)))\n",
    "    \n",
    "file = \"file_samples/PF00870_seed.sth\"\n",
    "outFile = \"file_samples/PF00870_seed.tsv\"\n",
    "\n",
    "infos = getStockholmInfo(file)\n",
    "print(\"\")\n",
    "writeStockholmInfo(outFile, infos)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "3. Load the contigs present in the [filtered_contigs.fasta](file_samples/filtered_contigs.fasta) file and translate each DNA sequence into the corresponding protein. Count the number of stop codons (i.e. \\*) for each sequence and print them to the user (e.g. MDC020656.85 51). Finally, write the translated proteins in another .fasta file (e.g. filtered_contigs_translated.fasta)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MDC001115.177 \t 65\n",
      "MDC013284.379 \t 93\n",
      "MDC018185.243 \t 418\n",
      "MDC018185.241 \t 467\n",
      "MDC004527.213 \t 48\n",
      "MDC012176.157 \t 30\n",
      "MDC001204.810 \t 155\n",
      "MDC004389.256 \t 43\n",
      "MDC018297.229 \t 317\n",
      "MDC001802.364 \t 45\n",
      "MDC014057.243 \t 23\n",
      "MDC021015.302 \t 19\n",
      "MDC017187.314 \t 83\n",
      "MDC012865.410 \t 68\n",
      "MDC000427.83 \t 42\n",
      "MDC017187.319 \t 251\n",
      "MDC004364.265 \t 145\n",
      "MDC002360.219 \t 60\n",
      "MDC015155.172 \t 47\n",
      "MDC019140.398 \t 67\n",
      "MDC019140.399 \t 56\n",
      "MDC011390.337 \t 71\n",
      "MDC007154.375 \t 248\n",
      "MDC010588.505 \t 29\n",
      "MDC002519.240 \t 500\n",
      "MDC006346.711 \t 376\n",
      "MDC011551.182 \t 806\n",
      "MDC002717.156 \t 55\n",
      "MDC006346.719 \t 157\n",
      "MDC007838.447 \t 141\n",
      "MDC007018.186 \t 47\n",
      "MDC017873.233 \t 35\n",
      "MDC016296.138 \t 176\n",
      "MDC019067.226 \t 49\n",
      "MDC014019.318 \t 45\n",
      "MDC007995.528 \t 53\n",
      "MDC026961.60 \t 71\n",
      "MDC020963.161 \t 65\n",
      "MDC005174.220 \t 132\n",
      "MDC040033.7 \t 142\n",
      "MDC019674.147 \t 89\n",
      "MDC010450.877 \t 40\n",
      "MDC007097.457 \t 89\n",
      "MDC016278.70 \t 69\n",
      "\n",
      "44 sequences written to output file\n"
     ]
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "\n",
    "seqIterator = SeqIO.parse(\"file_samples/filtered_contigs.fasta\", \"fasta\")\n",
    "\n",
    "sRcds = []\n",
    "for s in seqIterator:\n",
    "    #to avoid problems if not divisible by 3!\n",
    "    seq1 = s.seq.tomutable()\n",
    "    if(len(seq1) % 3 != 0):\n",
    "        while( len(seq1) % 3 != 0):\n",
    "            seq1 +=\"N\"\n",
    "    transl = seq1.toseq().translate()\n",
    "    stop_cnt = transl.count(\"*\")\n",
    "    print(s.id,\"\\t\", stop_cnt)\n",
    "    s.seq = transl\n",
    "    sRcds.append(s)\n",
    "\n",
    "N = SeqIO.write(sRcds,\"file_samples/filtered_contigs_translated.fasta\", \"fasta\")\n",
    "print(\"\")\n",
    "print(\"{} sequences written to output file\".format(N))\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "4. Write the following three functions:\n",
    "\n",
    "    1. ```countMatches(s1,s2)``` that takes two sequences aligned with pairwise2 local or global of the same length and returns how many elements match;\n",
    "    2. ```countMismatches(s1,s2)``` that takes two sequences aligned with pairwise2 local or global of the same length and returns how many elements are different (i.e. they are not gaps but the character is different);\n",
    "    3. ```countGapOpens(s1,s2)``` that takes two sequences aligned with pairwise2 local or global of the same length and returns the number of gap opened in the alignment;\n",
    "    4. ```countGapExtensions(s1,s2)``` that takes two sequences aligned with pairwise2 local or global of the same length and returns the number of gap extensions in the alignment\n",
    "    5. ```getScore(s1,s2, matchScore, mismatchPenalty, gapOpenPenalty, gapExtensionPenalty)``` and returns the score of the alignment of s1 and s2 (aligned with pairwise2 as before) given the scores and penalties in input.\n",
    "    \n",
    "Finally, align the sequences of the [Interleukin-12](https://en.wikipedia.org/wiki/Interleukin_12) chain A (let's call it s1) [IL12A.fasta](file_samples/IL12A.fasta) and B [IL12B.fasta](file_samples/IL12B.fasta) (let's call it s2) with the following parameters:\n",
    "\n",
    "    pairwise2.align.globalxx(s1,s2)\n",
    "    pairwise2.align.globalms(s1,s2,,1,-0.1,-1,-0.1)\n",
    "    pairwise2.align.globalxs(s1, s2,1,-0.5,-5,-0.5)\n",
    "  \n",
    "and check the score as computed from pairwise2 and from your functions (use the first of all the alignments to check). Note the effect of changing parameters.\n",
    "\n",
    "<div class=\"tggle\" onclick=\"toggleVisibility('ex4');\">Show/Hide Solution</div>\n",
    "<div id=\"ex4\" style=\"display:none;\">"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MCPAR------S---L--L---LVAT---L---VL----LDHLSL----ARNLP---VA--TP-D-P---GMFPC----LHHS-QNLLRAVSNM---LQ---KARQTL-----EF-----YPCTSE-----EID---H--------ED-I--T---KD-KTSTVEACLP-----L--ELT-KNE-S----C--LNSRETSF-I-TN----------GS-------CL-A---S--R-----KT----SFMMAL--CL---S-------S--IYE---D----LKMYQVE-----F------KTMNA----K-L-LMD-P-K--RQIFLDQNMLAVIDELMQALNFN-S-E---TV---PQK-S--SLE------------EP--D--FYKT-K-----IKLCILLHAFR----I--RAVTI-DRVMSYLN--------AS----\n",
      "MC---HQQLVISWFSLVFLASPLVA-IWELKKDV-YVVELD----WYPDA---PGEMV-VLT-CDTPEEDG----ITWTL---DQ------S--SEVL-GSGK---TLTIQVKEFGDAGQY--T--CHKGGE--VLSHSLLLLHKKEDGIWSTDILKDQK----E---PKNKTFLRCE--AKN-YSGRFTCWWL----T--TIST-DLTFSVKSSRGSSDPQGVTC-GAATLSAERVRGDNK-EYEYS-----VEC-QEDSACPAAEESLPI-EVMVDAVHKLK-Y--ENYTSSFFIRDIIK----PDPPKNLQL--KPLKNSR-----Q----V--E--------VSWEYPDT-WSTP--HSYFSL-TFCVQVQGKSKRE-KKDRVF--TDKTSATVI--C------RKNASISVRA---QDR---Y--YSSSWSEWASVPCS\n",
      "103.0\n",
      "\n",
      "Matches:103 Mismatches:0 GapOpens:88 GapExtensions:223\n",
      "Recomputed Score: 103\n",
      "\n",
      "MC------PARSL--L---LVA--TL------VLLDHLSLARNLPVA-TPD-P---GMFPCLHHSQNLLRAVSNM-LQKARQTLEF-----YPC------TS---------EE-IDHEDITKD-KTSTVEACLPLELTKNES----CLNSRETSFITNGSCLASRKTS-----FMMALCLS--SIYEDLK--MYQVEFKTMNA------------------KL---------------LMDP-KR-QIFLDQNMLAV------IDELMQALNFNSET----VPQKSSLEEPD--FY-KTKIKLCILLHAFRI--RAVTIDRVMS---YLNA----S\n",
      "MCHQQLVISWFSLVFLASPLVAIWELKKDVYVVELDWYPDAPGEMVVLTCDTPEEDGITWTLDQSSEVL--GSGKTL--TIQVKEFGDAGQYTCHKGGEVLSHSLLLLHKKEDGIWSTDILKDQKEPKNKTFLRCE-AKNYSGRFTCWWLTTISTDLTFSVKSSRGSSDPQGVTCGAATLSAERVRGDNKEYEYSVECQEDSACPAAEESLPIEVMVDAVHKLKYENYTSSFFIRDIIKPDPPKNLQLKPLKNSRQVEVSWEYPDTWSTPHSYFSLTFCVQVQGKSKREKKDRVFTDKT-SATVICRKNASISVRA--QDRYYSSSWSEWASVPCS\n",
      "28.19999999999986\n",
      "\n",
      "Matches:84 Mismatches:127 GapOpens:34 GapExtensions:91\n",
      "Recomputed Score: 28.199999999999996\n",
      "\n",
      "MCPARSLLLVATLVLLDHLSLA----------RNLPVATPDPGMFPCLHHSQNLLRAVSNMLQKARQTL----EFYPCTSEEIDHEDITKDKTSTVEACLPLELTKNESCLNS-------RETSFITNGSCLASRKTSFMMALCLSSIYEDL----------------------------------KMYQVEFKTMNA------------------KLLMDPKRQIFLDQNMLAVIDELMQALNFNSETVPQKSSLEEPDFYKT------------------------------------KIKLCILLHAFRIRAVTIDRVMSYLNAS\n",
      "MCHQQLVISWFSLVFLASPLVAIWELKKDVYVVELDWYPDAPGEMVVLTCDTPEEDGITWTLDQSSEVLGSGKTLTIQVKEFGDAGQYTCHKGGEVLSHSLLLLHKKEDGIWSTDILKDQKEPKNKTFLRCEAKNYSGRFTCWWLTTISTDLTFSVKSSRGSSDPQGVTCGAATLSAERVRGDNKEYEYSVECQEDSACPAAEESLPIEVMVDAVHKLKYENYTSSFFIRDIIKPDPPKNLQLKPLKNSRQVEVSWEYPDTWSTPHSYFSLTFCVQVQGKSKREKKDRVFTDKTSATVICRKNASISVRAQDRYYSSSWSEWASVPCS\n",
      "-122.0\n",
      "\n",
      "Matches:46 Mismatches:173 GapOpens:6 GapExtensions:103\n",
      "Recomputed Score: -122.0\n"
     ]
    }
   ],
   "source": [
    "from Bio import pairwise2\n",
    "from Bio import SeqIO\n",
    "\n",
    "\n",
    "def countMatches(a1,a2):\n",
    "    return sum([1 for i in range(len(a1)) if a1[i] == a2[i]])\n",
    "\n",
    "def countMismatches(a1,a2):\n",
    "    return sum([1 for i in range(len(a1)) if a1[i] != a2[i] and a1[i] != \"-\" and a2[i] != \"-\"])\n",
    "\n",
    "def countGapOpens(a1,a2):\n",
    "    gaps = [1 for i in range(1,len(a1)-1) \n",
    "            if ((a1[i] == \"-\" or a2[i] == \"-\") and a1[i-1] != \"-\" and a2[i-1] !=\"-\")\n",
    "           ]\n",
    "    return sum(gaps)\n",
    "\n",
    "def countGapExtensions(a1,a2):\n",
    "    gaps = [ 1 for i in range(1,len(a1)-1) \n",
    "            if ((a1[i] == \"-\" and a1[i-1] == \"-\") or (a2[i] == \"-\" and a2[i-1] ==\"-\"))\n",
    "           ]\n",
    "    return sum(gaps)\n",
    "\n",
    "def getScore(s1,s2, matchScore, mismatchPenalty, gapOpenPenalty, gapExtensionPenalty):\n",
    "    mat = countMatches(s1,s2)\n",
    "    mis = countMismatches(s1,s2)\n",
    "    gapo = countGapOpens(s1,s2)\n",
    "    gape = countGapExtensions(s1,s2)\n",
    "    score = matchScore*mat + mis*mismatchPenalty + gapo*gapOpenPenalty + gape*gapExtensionPenalty\n",
    "    return score\n",
    "\n",
    "def getSequenceData(filename):\n",
    "    record = SeqIO.read(filename, \"fasta\")\n",
    "    return record.seq\n",
    "\n",
    "def printInfo(a1,a2,computedScore,Mscore,mScore,goScore,geScore):\n",
    "    print(a1)\n",
    "    print(a2)\n",
    "    print(computedScore)\n",
    "    print(\"\")\n",
    "    M = countMatches(a1,a2)\n",
    "    m = countMismatches(a1,a2)\n",
    "    go = countGapOpens(a1,a2)\n",
    "    ge = countGapExtensions(a1,a2)\n",
    "    print(\"Matches:{} Mismatches:{} GapOpens:{} GapExtensions:{}\".format(M,m,go,ge))\n",
    "    print(\"Recomputed Score: {}\".format(getScore(a1,a2,Mscore,mScore,goScore,geScore)))\n",
    "\n",
    "s1 = getSequenceData(\"file_samples/IL12A.fasta\")\n",
    "s2 = getSequenceData(\"file_samples/IL12B.fasta\")\n",
    "\n",
    "aligns = pairwise2.align.globalxx(s1,s2)\n",
    "a1 = aligns[0][0]\n",
    "a2 = aligns[0][1]\n",
    "sc = aligns[0][2]\n",
    "printInfo(a1,a2,sc, 1,0,0,0)\n",
    "\n",
    "\n",
    "print(\"\")\n",
    "aligns = pairwise2.align.globalms(s1,s2,1,-0.1,-1,-0.1)\n",
    "a1 = aligns[0][0]\n",
    "a2 = aligns[0][1]\n",
    "sc = aligns[0][2]\n",
    "printInfo(a1,a2,sc, 1,-0.1,-1,-0.1)\n",
    "\n",
    "\n",
    "print(\"\")\n",
    "aligns = pairwise2.align.globalms(s1, s2,1,-0.5,-5,-0.5)\n",
    "\n",
    "a1 = aligns[0][0]\n",
    "a2 = aligns[0][1]\n",
    "sc = aligns[0][2]\n",
    "printInfo(a1,a2,sc, 1,-0.5,-5,-0.5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "</div>"
   ]
  }
 ],
 "metadata": {
  "celltoolbar": "Edit Metadata",
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
