{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "nbsphinx": "hidden"
   },
   "outputs": [],
   "source": [
    "import jupman\n",
    "jupman.init()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Practical 7\n",
    "\n",
    "In this practical we will keep practicing with functions and will see how to get input from the command line. \n",
    "\n",
    "## Slides\n",
    "\n",
    "The slides of the introduction can be found here: [Intro](docs/Practical7.pdf)\n",
    "\n",
    "## Functions\n",
    "\n",
    "\n",
    "Reminder. The basic definition of a function is:\n",
    "```\n",
    "def function_name(input) :\n",
    "    #code implementing the function\n",
    "    ...\n",
    "    ...\n",
    "    return return_value\n",
    "```\n",
    "\n",
    "Functions are defined with the **def** keyword that proceeds the *function_name* and then a list of parameters is passed in the brackets. A colon **:** is used to end the line holding the definition of the function. The code implementing the function is specified by using indentation. A function **might** or **might not** return a value. In the first case a **return** statement is used.\n",
    "\n",
    "\n",
    "## Getting input from the command line\n",
    "\n",
    "To call a program ```my_python_program.py``` from command line, you just have to open a terminal (in Linux) or the command prompt (in Windows) and, assuming that python is present in the path, you can ```cd``` into the folder containing your python program, (eg. ```cd C:\\python\\my_exercises\\```) and just type in \n",
    "```python3 my_python_program.py```\n",
    "or\n",
    "```python my_python_program.py```\n",
    "In case of arguments to be passed by command line, one has to put them after the specification of the program name (eg. ```python my_python_program.py parm1 param2 param3```\n",
    "\n",
    "Python provides the module **sys** to interact with the interpreter. In particular, **sys.argv** is a list representing all the arguments passed to the python script from the command line.\n",
    "\n",
    "Consider the following code:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "\"\"\"Test input from command line in systest.py\"\"\"\n",
    "\n",
    "if(len(sys.argv) != 4):\n",
    "    print(\"Dear user, I was expecting 3 params. You gave me \",len(sys.argv)-1)\n",
    "    sys.exit(1)\n",
    "else:\n",
    "    for i in range(0,len(sys.argv)):\n",
    "        print(\"Param {}:{} ({})\".format(i,sys.argv[i],type(sys.argv[i])))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Invoking the ```systest.py``` script from command line with the command  ```python3 exercises/systest.py 1st_param 2nd 3``` will return:\n",
    "```\n",
    "Param 0: exercises/systest.py (<class 'str'>)\n",
    "Param 1: 1st_param (<class 'str'>)\n",
    "Param 2: 2nd (<class 'str'>)\n",
    "Param 3: 3 (<class 'str'>)\n",
    "```\n",
    "Invoking the ```systest.py``` script from command line with the command  ```python3 exercises/systest.py 1st_param``` will return:\n",
    "```\n",
    "Dear user, I was expecting three parameters. You gave me  1\n",
    "```\n",
    "\n",
    "Note that the parameter at index 0, ```sys.argv[0]``` holds the name of the script, and that all parameters are actually **strings** (and therefore need to be cast to numbers if we want to do mathematical operations on them)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "A more flexible and powerful way of getting input from command line makes use of the ```Argparse``` [module](https://docs.python.org/3/howto/argparse.html). \n",
    "\n",
    "## Argparse\n",
    "\n",
    "Argparse is a command line parsing module which deals with user specified parameters (positional arguments) and optional arguments.\n",
    "\n",
    "\n",
    "Very briefly, the basic syntax of the ```Argparse module``` (for more information check the [official documentation](https://docs.python.org/3/howto/argparse.html)) is the following.\n",
    "\n",
    "1. Import the module:\n",
    "\n",
    "```\n",
    "import argparse\n",
    "```\n",
    "\n",
    "2. Define a argparse object:\n",
    "\n",
    "```\n",
    "parser = argparse.ArgumentParser(description=\"This is the description of the program\")\n",
    "```\n",
    "\n",
    "note the parameter *description* that is a string to describe the program;\n",
    "\n",
    "3. Add positional arguments:\n",
    "```\n",
    "parser.add_argument(\"arg_name\", type = obj, \n",
    "                    help = \"Description of the parameter)\n",
    "```\n",
    "where ```arg_name``` is the name of the argument (which will be used to retrieve its value). The argument has type ```obj``` (the type will be automatically checked for us) and a description specified in the ```help```string.\n",
    "\n",
    "4. Add optional arguments:\n",
    "```\n",
    "parser.add_argument(\"-p\", \"--positional_arg\", type = obj, default = def_val, \n",
    "                        help = \"Description of the parameter)\n",
    "```\n",
    "where ```-p``` is a short form of the parameter (and it is optional), ```--positional_arg``` is the extended name and it requires a value after it is specified, ```default``` is optional and gives a default value to the parameter. If not specified and no argument is passed, the argument will get the value \"None\". ```Help``` is again the description string.\n",
    "\n",
    "5. Parse the arguments:\n",
    "```\n",
    "args = parser.parse_args()\n",
    "```\n",
    "the parser checks the arguments and stores their values in the ```argparse``` object that we called ```args```.\n",
    "\n",
    "6. Retrieve and process arguments:\n",
    "```\n",
    "myArgName = args.arg_name\n",
    "myPosArg = args.positional_arg\n",
    "```\n",
    "now variables contain the values specified by the user and we can use them.\n",
    "\n",
    "**Example:**\n",
    "Let's write a program that gets a string (S) and an integer (N) in input and prints the string repeated N times. Three optional parameters are specified: verbosity (-v) to make the software print a more descriptive output, separator (-s) to separate each copy of the string (defaults to \" \") and trailpoints (-p) to add several \".\" at the end of the string (defaults to 1). "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import argparse\n",
    "parser = argparse.ArgumentParser(description=\"\"\"This script gets a string \n",
    "                                 and an integer and repeats the string N times\"\"\")\n",
    "parser.add_argument(\"string\", type=str,\n",
    "                    help=\"The string to be repeated\")\n",
    "parser.add_argument(\"N\", type=int,\n",
    "                    help=\"The number of time to repeated the string\")\n",
    "\n",
    "parser.add_argument(\"-v\", \"--verbose\", action=\"store_true\",\n",
    "                    help=\"increase output verbosity\")\n",
    "\n",
    "parser.add_argument(\"-p\", \"--trailpoints\", type = int, default = 1, \n",
    "                    help=\"Adds these many trailing points\")\n",
    "parser.add_argument(\"-s\", \"--separator\", type = str, default = \" \", \n",
    "                    help=\"The separator between repeated strings\")\n",
    "\n",
    "args = parser.parse_args()\n",
    "\n",
    "mySTR = args.string+args.separator\n",
    "trailP = \".\" * args.trailpoints\n",
    "answer = mySTR * args.N \n",
    "\n",
    "answer = answer[:-len(args.separator)] + trailP #to remove the last separator\n",
    "\n",
    "if args.verbose:\n",
    "    print(\"the string {} repeated {} is:\".format(args.str, args.N, answer))\n",
    "else:\n",
    "    print(answer)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Executing the program from command line without parameters gives the message:\n",
    "\n",
    "![](img/pract7/noargs.png)\n",
    "\n",
    "Calling it with the ```-h``` flag:\n",
    "\n",
    "![](img/pract7/help.png)\n",
    "\n",
    "With the positional arguments ```\"ciao a tutti\"``` and ```3```:\n",
    "\n",
    "![](img/pract7/pos_args.png)\n",
    "\n",
    "With the positional arguments ```\"ciao a tutti\"``` and ```3```, and with the optional parameters ```-s \"___\" -p 3 -v```\n",
    "\n",
    "![](img/pract7/sample.png)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Example:**\n",
    "Let's write a program that reads and prints to screen a text file specified by the user. Optionally, the file might be compressed with gzip to save space. The user should be able to read also gzipped files. Hint: use the module gzip which is very similar to the standard file management method ([more info here](https://docs.python.org/3/library/gzip.html?highlight=gzip#module-gzip)). You can file a text file here [textFile.txt](file_samples/textFile.txt) and its gzipped version here [text.gz](file_samples/textFile.gz):\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import argparse\n",
    "import gzip\n",
    "\n",
    "parser = argparse.ArgumentParser(description=\"\"\"Reads and prints a text file\"\"\")\n",
    "parser.add_argument(\"filename\", type=str, help=\"The file name\")\n",
    "parser.add_argument(\"-z\", \"--gzipped\", action=\"store_true\", help=\"If set, input file is assumed gzipped\")\n",
    "\n",
    "args = parser.parse_args()\n",
    "inputFile = args.filename\n",
    "fh = \"\"\n",
    "if(args.gzipped):\n",
    "    fh = gzip.open(inputFile, \"rt\")\n",
    "else:\n",
    "    fh = open(inputFile, \"r\")\n",
    "\n",
    "for line in fh:\n",
    "    line = line.strip(\"\\n\")\n",
    "    print(line)\n",
    "\n",
    "fh.close()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The output:\n",
    "\n",
    "![](img/pract7/read_gz.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Exercises\n",
    "\n",
    "1. Modify the program of Exercise 4 of Practical 6 in order to allow users to specify the input and output files from command line. Then test it with the provided files. The text of the exercise follows:\n",
    "\n",
    "Write a python program that reads two files. The first is a one column text file ([contig_ids.txt](file_samples/contig_ids.txt)) with the identifiers of some contigs that are present in the second file, which is a fasta formatted file ([contigs82.fasta](file_samples/contigs82.fasta)). The program will write on a third, fasta formatted file (e.g. filtered_contigs.fasta) only those entries in *contigs82.fasta* having identifier in *contig_ids.txt*.\n",
    "\n",
    "\n",
    "\n",
    "<div class=\"tggle\" onclick=\"toggleVisibility('ex1');\">Show/Hide Solution</div>\n",
    "<div id=\"ex1\" style=\"display:none;\">"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "import argparse\n",
    "\n",
    "def readIDS(f):\n",
    "    \"\"\"reads a one column file in and stores\n",
    "    the ids in a dictionary that is returned at the end\"\"\"\n",
    "    ret = dict()\n",
    "    with open(f, \"r\") as file:\n",
    "        for line in file:\n",
    "            line = line.strip()\n",
    "            if(line not in ret):\n",
    "                ret[line] = 1\n",
    "    return ret\n",
    "\n",
    "def filterFasta(inF, outF, ids2keep):\n",
    "    oF = open(outF, \"w\")\n",
    "    \n",
    "    outputME = False\n",
    "    with open(inF, \"r\") as file:\n",
    "        for line in file:\n",
    "            line = line.strip()\n",
    "            if(line.startswith(\">\")):\n",
    "                #this is the header\n",
    "                if(line[1:] in ids2keep):\n",
    "                    oF.write(line +\"\\n\")\n",
    "                    outputME = True\n",
    "                    print(\"Writing contig \", line[1:])\n",
    "                else:\n",
    "                    outputME = False\n",
    "            else:\n",
    "                if(outputME):\n",
    "                    oF.write(line +\"\\n\")\n",
    "        \n",
    "    oF.close()\n",
    "    \n",
    "\n",
    "parser = argparse.ArgumentParser(description=\"Filters a fasta file\")\n",
    "parser.add_argument(\"inputFasta\", type = str, help = \"The input fasta file\")\n",
    "parser.add_argument(\"inputIDS\", type = str, help = \"The IDS to keep\")\n",
    "parser.add_argument(\"outputFasta\", type = str, help = \"The output fasta file with filtered entries\")\n",
    "args = parser.parse_args()\n",
    "idsFile = args.inputIDS\n",
    "inFasta = args.inputFasta\n",
    "outFasta = args.outputFasta\n",
    "\n",
    "ids = readIDS(idsFile)\n",
    "filterFasta(inFasta,outFasta, ids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "2. [Cytoscape](https://www.cytoscape.org) is a well known tool to perform network analysis. It is well integrated with several online databases housing for example protein-protein interactions like EBI's [IntAct](https://www.ebi.ac.uk/intact/). It is also able to read and write a very simple text file called ```.sif``` to represent interactions between the nodes of a network. Sif formatted files are tab separated (```\\t```) and each line represents a connection between the nodes of the network. For example:\n",
    "```\n",
    "node1 interaction1 node2\n",
    "node1 interaction2 node3\n",
    "node2 interaction1 node3\n",
    "```\n",
    "represents two types of interactions between node1, node2 and node3. Normally nodes are represented as circles in a network (graph) and interactions as lines (that can be of different kinds) connecting nodes (edges). The following is an extract from the file [pka.sif](file_samples/pka.sif) that has been downloaded by Cytoscape from the database IntAct and represents the interactions of the Protein Kinase A (PKA) of E.coli:\n",
    "```\n",
    "P75742\tEBI-9168813\tP76594\n",
    "P21513\tEBI-888473\tP76594\n",
    "P21513\tEBI-15543881\tP76594\n",
    "```\n",
    "the first and third columns represent proteins and the second is the interaction joining them. All the values are identifiers from the IntAct database. The cytoscape representation of the full set of interactions is:\n",
    "\n",
    "![](img/pract7/pka_ppi.png)\n",
    "\n",
    "Write a python script that reads in the .sif file ([pka.sif](file_samples/pka.sif) is here but even better if any .sif file specified in input by the user) and stores the information in one (or more) suitable objects to be able to:\n",
    "    \n",
    "    1. Print the interaction that is more present among the nodes;\n",
    "    2. Print the node that is connected to the highest number of other nodes (no matter if on the left or right of the interaction);\n",
    "    \n",
    "\n",
    "**Hint:** you can store the information in a dictionary having the interaction as key and a list of tuples (node1,node2) as value. Although redundant, it is convenient to keep a list of unique nodes. \n",
    "**Note:** This will use more memory but it is acceptable for small examples as it allows to quickly answer the questions.\n",
    "\n",
    "Optional: check what these are on the [IntAct database](https://www.ebi.ac.uk/intact/).\n",
    "    \n",
    "<div class=\"tggle\" onclick=\"toggleVisibility('ex2');\">Show/Hide Solution</div>\n",
    "<div id=\"ex2\" style=\"display:none;\">"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The most present interaction(s): EBI-891722,EBI-15548579. Present 22 times\n",
      "The most present node(s) P76594. Present 56 times\n"
     ]
    }
   ],
   "source": [
    "import argparse\n",
    "\n",
    "def readSif(fn):\n",
    "    \"\"\"reads in the sif file fn and ouptputs a dictionary and a list:\n",
    "    interDict : has \"interaction\" has keys and [(node1,node2), (node3,node4)] as values \n",
    "    node list : is the list of unique node names (both on the right or left hand of an interaction)\n",
    "    \"\"\"\n",
    "    interDict = {}\n",
    "    nodes = []\n",
    "    with open(fn, \"r\") as myfile:\n",
    "        for line in myfile:\n",
    "            line = line.strip()\n",
    "            [n1,inter,n2] = line.split(\"\\t\")\n",
    "            if inter not in interDict:\n",
    "                interDict[inter] = [(n1,n2)]\n",
    "            else:\n",
    "                interDict[inter].append((n1,n2))\n",
    "            \n",
    "            if n1 not in nodes:\n",
    "                nodes.append(n1)\n",
    "            if n2 not in nodes:\n",
    "                nodes.append(n2)\n",
    "    \n",
    "    return interDict,nodes\n",
    "\n",
    "def getMostPresentInteraction(iDict):\n",
    "    \"\"\"gets the interaction dictionary as defined above and returns the\n",
    "    most present interaction (with its count). If more than one, all are\n",
    "    returned comma separated\"\"\"\n",
    "    mpInter = \"\"\n",
    "    mpInterCount = 0\n",
    "    for inter in iDict:\n",
    "        \n",
    "        cnt = len(iDict[inter])\n",
    "        \n",
    "        if mpInterCount < cnt:\n",
    "            mpInterCount = cnt\n",
    "            mpInter = inter\n",
    "        elif mpInterCount == cnt:\n",
    "            mpInterCount = cnt\n",
    "            mpInter = mpInter + \",\" + inter\n",
    "    return (mpInter,mpInterCount)\n",
    "\n",
    "def getMostPresentNode(iDict, nodeList):\n",
    "    \"\"\"gets the most highly connected node (or nodes, comma separated)\n",
    "    and returns it with its count\n",
    "    iDict : the interaction dictionary seen above\n",
    "    nodeList : the node of unique node names seen above\n",
    "    \"\"\"\n",
    "    mostPresentNode = \"\"\n",
    "    mostPresentCount = 0\n",
    "    #NOTE: iDict : {\"inter1\" : [(n1,n2), (n2,n3),(n3,n1), (n1,n3)], \n",
    "    #               \"inter2\" : [(n5,n1), (n1,n2)] }\n",
    "    #\n",
    "    for n in nodeList:\n",
    "        #the number of times a node is present is the sum of its count in all interactions\n",
    "        #both as first or second member of the couple\n",
    "        curCnt = 0\n",
    "        for i in iDict:\n",
    "            pairsContainingN = [x for x in iDict[i] if n in x]\n",
    "            curCnt = curCnt + len(pairsContainingN)\n",
    "        if(curCnt > mostPresentCount):\n",
    "            mostPresentNode = n\n",
    "            mostPresentCount = curCnt\n",
    "        elif curCnt == mostPresentCount:\n",
    "            postPresentNode += \",\" + n\n",
    "            mostPresentCount = curCnt\n",
    "    \n",
    "    return (mostPresentNode,mostPresentCount)\n",
    "            \n",
    "#parser = argparse.ArgumentParser(description=\"\"\"Reads and processes a .sif file\"\"\")\n",
    "#parser.add_argument(\"filename\", type=str, help=\"The .sif file name\")\n",
    "\n",
    "#args = parser.parse_args()\n",
    "#inputFile = args.filename\n",
    "inputFile = \"file_samples/pka.sif\"\n",
    "interactions, nodeL = readSif(inputFile)\n",
    "(mpI,mpICount) = getMostPresentInteraction(interactions)\n",
    "print(\"The most present interaction(s): {}. Present {} times\".format(mpI,mpICount))\n",
    "(mpN, mpNCount) = getMostPresentNode(interactions,nodeL)\n",
    "print(\"The most present node(s) {}. Present {} times\".format(mpN,mpNCount))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "3. Given a [fasta](https://en.wikipedia.org/wiki/FASTA_format) file like [contigs82.fasta](file_samples/contigs82.fasta) specified in input by a user,  write a python script that counts, for each sequence, the number of times that a DNA or protein string specified in input appears. Print to the screen the result like:\n",
    "```\n",
    "contig_id124 contains sequence \"TGCTCCGAGCATGTGATAATCATTCCAAGCTCCAT\" 2 times\n",
    "```\n",
    "\n",
    "Modify the program so that it outputs also the list of all the indexes where the string appears in each sequence in the fasta file. Try to look for the following sequences:\n",
    "```\n",
    "TTTTCCTAGG\n",
    "TGCTCCGAGCATGTGATAATCATTCCAAGCTCCAT\n",
    "TAAACAT\n",
    "GATTACA\n",
    "```\n",
    "\n",
    "\n",
    "\n",
    "<div class=\"tggle\" onclick=\"toggleVisibility('ex3');\">Show/Hide Solution</div>\n",
    "<div id=\"ex3\" style=\"display:none;\">"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TAAACAT in MDC020656.85: 1 times ([2546])\n",
      "TAAACAT in MDC018185.243: 3 times ([1793, 19581, 22638])\n",
      "TAAACAT in MDC018185.241: 5 times ([2003, 7491, 19111, 21020, 21333])\n",
      "TAAACAT in MDC001204.812: 1 times ([1733])\n",
      "TAAACAT in MDC001204.810: 1 times ([740])\n",
      "TAAACAT in MDC018297.229: 1 times ([5483])\n",
      "TAAACAT in MDC001802.364: 1 times ([1018])\n",
      "TAAACAT in MDC018185.242: 7 times ([5186, 7704, 7804, 8087, 8733, 10603, 12892])\n",
      "TAAACAT in MDC017187.314: 1 times ([4144])\n",
      "TAAACAT in MDC000427.83: 3 times ([494, 549, 2189])\n",
      "TAAACAT in MDC017187.319: 2 times ([3564, 9145])\n",
      "TAAACAT in MDC017187.318: 1 times ([1876])\n",
      "TAAACAT in MDC015147.205: 2 times ([2145, 4161])\n",
      "TAAACAT in MDC026201.7: 1 times ([419])\n",
      "TAAACAT in MDC003995.601: 1 times ([8163])\n",
      "TAAACAT in MDC054294.001: 6 times ([7820, 10001, 10558, 13088, 17674, 20102])\n",
      "TAAACAT in MDC003408.117: 3 times ([2659, 2695, 2780])\n",
      "TAAACAT in MDC053310.000: 2 times ([961, 1820])\n",
      "TAAACAT in MDC019140.398: 2 times ([128, 2196])\n",
      "TAAACAT in MDC002519.240: 3 times ([13449, 14983, 26536])\n",
      "TAAACAT in MDC006346.711: 5 times ([3103, 5760, 5865, 8452, 17240])\n",
      "TAAACAT in MDC011551.182: 5 times ([17811, 20474, 23891, 25280, 35173])\n",
      "TAAACAT in MDC006346.719: 1 times ([2520])\n",
      "TAAACAT in MDC007838.447: 1 times ([1023])\n",
      "TAAACAT in MDC016296.138: 1 times ([2063])\n",
      "TAAACAT in MDC019067.226: 2 times ([1188, 1543])\n",
      "TAAACAT in MDC036568.1: 1 times ([1331])\n",
      "TAAACAT in MDC022800.298: 1 times ([660])\n",
      "TAAACAT in MDC002479.192: 1 times ([654])\n",
      "TAAACAT in MDC020963.162: 1 times ([12853])\n",
      "TAAACAT in MDC051637.000: 2 times ([1560, 10976])\n"
     ]
    }
   ],
   "source": [
    "import argparse\n",
    "\n",
    "def getIndexes(string1,string2):\n",
    "    \"\"\"checks if string2 is present in string1 and returns \n",
    "    all the positions at which string2 occurs in string1\"\"\"\n",
    "    ret = []\n",
    "    ind = string1.find(string2)\n",
    "    \n",
    "    while (ind > -1 and ind < len(string1)):\n",
    "        ret.append(ind)\n",
    "        ind = string1.find(string2,ind + 1)\n",
    "        \n",
    "    return ret\n",
    "\n",
    "def processFasta(file,testStr):\n",
    "    \"\"\"reads a fasta file entry by entry checks if the input string\n",
    "       testStr is present in each sequence. Reporting how many times and where.\n",
    "    \"\"\"\n",
    "    header = \"\"\n",
    "    seq = \"\"\n",
    "    with open(file, \"r\") as f:\n",
    "        for line in f:\n",
    "            line = line.strip()\n",
    "            if(line.startswith(\">\")):\n",
    "                if(len(header) == 0 ):\n",
    "                    #first entry:\n",
    "                    header = line[1:]\n",
    "                else:\n",
    "                    #this is a new entry\n",
    "                    indexes = getIndexes(seq,testStr)\n",
    "                    if len(indexes) > 0:\n",
    "                        print(\"{} in {}: {} times ({})\".format(testStr, header,len(indexes),indexes))\n",
    "                    seq = \"\"\n",
    "                    header = line[1:]\n",
    "            else:\n",
    "                seq +=line\n",
    "    #processing the final entry\n",
    "    indexes = getIndexes(seq,testStr)\n",
    "    if len(indexes) > 0:\n",
    "        print(\"{} in {}: {} times ({})\".format(testStr, header,len(indexes),indexes))\n",
    "                    \n",
    "parser = argparse.ArgumentParser(description=\"\"\"Checks if a sequence is exactly contained in a fasta file\"\"\")\n",
    "parser.add_argument(\"filename\", type=str, help=\"The fasta file name\")\n",
    "parser.add_argument(\"query\", type=str, help=\"The query string\")\n",
    "\n",
    "args = parser.parse_args()\n",
    "inFasta = args.filename\n",
    "testS = args.query\n",
    "#inFasta = \"file_samples/contigs82.fasta\"\n",
    "#testS = \"TAAACAT\"\n",
    "processFasta(inFasta, testS)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "4. The [Fisher's dataset](http://onlinelibrary.wiley.com/doi/10.1111/j.1469-1809.1936.tb02137.x/abstract) regarding Petal and Sepal length and width in csv format can be found [here](file_samples/Fishers_Iris.csv). These are the measurements of  the flowers of  fifty plants each of the two species Iris setosa and Iris versicolor.\n",
    "\n",
    "The header of the file is:\n",
    "```\n",
    "Species Number,Species Name,Petal width,Petal length,Sepal length,Sepal width\n",
    "```\n",
    "\n",
    "Write a python script that reads this file in input (feel free to hard-code the filename in the code) and computes the average petal length and width and sepal length and width for each of the three different Iris species. Print them to the screen alongside the number of elements.\n",
    "<div class=\"tggle\" onclick=\"toggleVisibility('ex4');\">Show/Hide Solution</div>\n",
    "<div id=\"ex4\" style=\"display:none;\">"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Species I. Setosa has 50 measurements:\n",
      "\t petal length 14.62\n",
      "\t petal width 2.46\n",
      "\t sepal length 34.28\n",
      "\t sepal width 50.1\n",
      "Species I. Versicolor has 50 measurements:\n",
      "\t petal length 43.22\n",
      "\t petal width 13.26\n",
      "\t sepal length 27.64\n",
      "\t sepal width 59.36\n",
      "Species I. Verginica has 50 measurements:\n",
      "\t petal length 55.52\n",
      "\t petal width 20.06\n",
      "\t sepal length 29.74\n",
      "\t sepal width 65.88\n"
     ]
    }
   ],
   "source": [
    "def readCSV(f):\n",
    "    \"\"\"reads the csv dataset and returns a dictionary with \n",
    "    species name as key and, as value, a dictionary with four keys : petalLen, sepalLen, petalWidth, sepalWidth\n",
    "    \"\"\"\n",
    "    ret = dict()\n",
    "    with open(f, \"r\") as file:\n",
    "        for line in file:\n",
    "            line = line.strip()\n",
    "            if not line.startswith(\"Species Number\"):\n",
    "                data = line.split(\",\")\n",
    "                speciesName = data[1]\n",
    "                pWidth = int(data[2])\n",
    "                pLen = int(data[3])\n",
    "                sLen = int(data[4])\n",
    "                sWidth = int(data[5])\n",
    "                if(speciesName not in ret):\n",
    "                    ret[speciesName] = {\"petalLen\" : [], \"sepalLen\" : [], \n",
    "                                        \"petalWidth\" : [], \"sepalWidth\" : []\n",
    "                                       }\n",
    "                ret[speciesName][\"petalLen\"].append(pLen)\n",
    "                ret[speciesName][\"sepalLen\"].append(sLen)\n",
    "                ret[speciesName][\"sepalWidth\"].append(sWidth)\n",
    "                ret[speciesName][\"petalWidth\"].append(pWidth)\n",
    "    return ret\n",
    "\n",
    "def printData(dataDict):\n",
    "    for s in dataDict:\n",
    "        avgPlen = sum(dataDict[s][\"petalLen\"])/len(dataDict[s][\"petalLen\"])\n",
    "        avgPwid = sum(dataDict[s][\"petalWidth\"])/len(dataDict[s][\"petalWidth\"])\n",
    "        avgSlen = sum(dataDict[s][\"sepalLen\"])/len(dataDict[s][\"sepalLen\"])\n",
    "        avgSwid = sum(dataDict[s][\"sepalWidth\"])/len(dataDict[s][\"sepalWidth\"])\n",
    "        print(\"Species {} has {} measurements:\".format(s, len(dataDict[s][\"petalLen\"])))\n",
    "        print(\"\\t petal length {}\".format(avgPlen))\n",
    "        print(\"\\t petal width {}\".format(avgPwid))\n",
    "        print(\"\\t sepal length {}\".format(avgSlen))\n",
    "        print(\"\\t sepal width {}\".format(avgSwid))\n",
    "        \n",
    "        \n",
    "        \n",
    "\n",
    "inFile = \"file_samples/Fishers_Iris.csv\"\n",
    "\n",
    "data = readCSV(inFile)\n",
    "printData(data)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "</div>"
   ]
  }
 ],
 "metadata": {
  "celltoolbar": "Edit Metadata",
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
