{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "nbsphinx": "hidden"
   },
   "outputs": [],
   "source": [
    "import jupman\n",
    "jupman.init()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Practical 7\n",
    "\n",
    "In this practical we will keep practicing with functions and will see how to get input from the command line. \n",
    "\n",
    "## Slides\n",
    "\n",
    "The slides of the introduction can be found here: [Intro](docs/Practical7.pdf)\n",
    "\n",
    "## Functions\n",
    "\n",
    "\n",
    "Reminder. The basic definition of a function is:\n",
    "```\n",
    "def function_name(input) :\n",
    "    #code implementing the function\n",
    "    ...\n",
    "    ...\n",
    "    return return_value\n",
    "```\n",
    "\n",
    "Functions are defined with the **def** keyword that proceeds the *function_name* and then a list of parameters is passed in the brackets. A colon **:** is used to end the line holding the definition of the function. The code implementing the function is specified by using indentation. A function **might** or **might not** return a value. In the first case a **return** statement is used.\n",
    "\n",
    "\n",
    "## Getting input from the command line\n",
    "\n",
    "To call a program ```my_python_program.py``` from command line, you just have to open a terminal (in Linux) or the command prompt (in Windows) and, assuming that python is present in the path, you can ```cd``` into the folder containing your python program, (eg. ```cd C:\\python\\my_exercises\\```) and just type in \n",
    "```python3 my_python_program.py```\n",
    "or\n",
    "```python my_python_program.py```\n",
    "In case of arguments to be passed by command line, one has to put them after the specification of the program name (eg. ```python my_python_program.py parm1 param2 param3```\n",
    "\n",
    "Python provides the module **sys** to interact with the interpreter. In particular, **sys.argv** is a list representing all the arguments passed to the python script from the command line.\n",
    "\n",
    "Consider the following code:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "\"\"\"Test input from command line in systest.py\"\"\"\n",
    "\n",
    "if(len(sys.argv) != 4):\n",
    "    print(\"Dear user, I was expecting three parameters. You gave me \",len(sys.argv)-1)\n",
    "    sys.exit(1)\n",
    "else:\n",
    "    for i in range(0,len(sys.argv)):\n",
    "        print(\"Param {}: {} ({})\".format(i,sys.argv[i], type(sys.argv[i])))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Invoking the ```systest.py``` script from command line with the command  ```python3 exercises/systest.py 1st_param 2nd 3``` will return:\n",
    "```\n",
    "Param 0: exercises/systest.py (<class 'str'>)\n",
    "Param 1: 1st_param (<class 'str'>)\n",
    "Param 2: 2nd (<class 'str'>)\n",
    "Param 3: 3 (<class 'str'>)\n",
    "```\n",
    "Invoking the ```systest.py``` script from command line with the command  ```python3 exercises/systest.py 1st_param``` will return:\n",
    "```\n",
    "Dear user, I was expecting three parameters. You gave me  1\n",
    "```\n",
    "\n",
    "Note that the parameter at index 0, ```sys.argv[0]``` holds the name of the script, and that all parameters are actually **strings** (and therefore need to be cast to numbers if we want to do mathematical operations on them)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "A more flexible and powerful way of getting input from command line makes use of the ```Argparse``` [module](https://docs.python.org/3/howto/argparse.html). \n",
    "\n",
    "## Argparse\n",
    "\n",
    "Argparse is a command line parsing module which deals with user specified parameters (positional arguments) and optional arguments.\n",
    "\n",
    "\n",
    "Very briefly, the basic syntax of the ```Argparse module``` (for more information check the [official documentation](https://docs.python.org/3/howto/argparse.html)) is the following.\n",
    "\n",
    "1. Import the module:\n",
    "\n",
    "```\n",
    "import argparse\n",
    "```\n",
    "\n",
    "2. Define a argparse object:\n",
    "\n",
    "```\n",
    "parser = argparse.ArgumentParser(description=\"This is the description of the program\")\n",
    "```\n",
    "\n",
    "note the parameter *description* that is a string to describe the program;\n",
    "\n",
    "3. Add positional arguments:\n",
    "```\n",
    "parser.add_argument(\"arg_name\", type = obj, help = \"Description of the parameter)\n",
    "```\n",
    "where ```arg_name``` is the name of the argument (which will be used to retrieve its value). The argument has type ```obj``` (the type will be automatically checked for us) and a description specified in the ```help```string.\n",
    "\n",
    "4. Add optional arguments:\n",
    "```\n",
    "parser.add_argument(\"-p\", \"--positional_arg\", type = obj, default = def_val, help = \"Description of the parameter)\n",
    "```\n",
    "where ```-p``` is a short form of the parameter (and it is optional), ```--positional_arg``` is the extended name and it requires a value after it is specified, ```default``` is optional and gives a default value to the parameter. If not specified and no argument is passed, the argument will get the value \"None\". ```Help``` is again the description string.\n",
    "\n",
    "5. Parse the arguments:\n",
    "```\n",
    "args = parser.parse_args()\n",
    "```\n",
    "the parser checks the arguments and stores their values in the ```argparse``` object that we called ```args```.\n",
    "\n",
    "6. Retrieve and process arguments:\n",
    "```\n",
    "myArgName = args.arg_name\n",
    "myPosArg = args.positional_arg\n",
    "```\n",
    "now variables contain the values specified by the user and we can use them.\n",
    "\n",
    "**Example:**\n",
    "Let's write a program that gets a string (S) and an integer (N) in input and prints the string repeated N times. Three optional parameters are specified: verbosity (-v) to make the software print a more descriptive output, separator (-s) to separate each copy of the string (defaults to \" \") and trailpoints (-p) to add several \".\" at the end of the string (defaults to 1). "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import argparse\n",
    "parser = argparse.ArgumentParser(description=\"\"\"This script gets a string \n",
    "                                 and an integer and repeats the string N times\"\"\")\n",
    "parser.add_argument(\"string\", type=str,\n",
    "                    help=\"The string to be repeated\")\n",
    "parser.add_argument(\"N\", type=int,\n",
    "                    help=\"The number of time to repeated the string\")\n",
    "\n",
    "parser.add_argument(\"-v\", \"--verbose\", action=\"store_true\",\n",
    "                    help=\"increase output verbosity\")\n",
    "\n",
    "parser.add_argument(\"-p\", \"--trailpoints\", type = int, default = 1, help=\"Adds these many trailing points\")\n",
    "parser.add_argument(\"-s\", \"--separator\", type = str, default = \" \", help=\"The separator between repeated strings\")\n",
    "\n",
    "args = parser.parse_args()\n",
    "\n",
    "mySTR = args.string+args.separator\n",
    "trailP = \".\" * args.trailpoints\n",
    "answer = mySTR * args.N \n",
    "\n",
    "answer = answer[:-len(args.separator)] + trailP #to remove the last separator\n",
    "\n",
    "if args.verbose:\n",
    "    print(\"the string {} repeated {} is:\".format(args.str, args.N, answer))\n",
    "else:\n",
    "    print(answer)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Executing the program from command line without parameters gives the message:\n",
    "\n",
    "![](img/pract7/noargs.png)\n",
    "\n",
    "Calling it with the ```-h``` flag:\n",
    "\n",
    "![](img/pract7/help.png)\n",
    "\n",
    "With the positional arguments ```\"ciao a tutti\"``` and ```3```:\n",
    "\n",
    "![](img/pract7/pos_args.png)\n",
    "\n",
    "With the positional arguments ```\"ciao a tutti\"``` and ```3```, and with the optional parameters ```-s \"___\" -p 3 -v```\n",
    "\n",
    "![](img/pract7/sample.png)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Example:**\n",
    "Let's write a program that reads and prints to screen a text file specified by the user. Optionally, the file might be compressed with gzip to save space. The user should be able to read also gzipped files. Hint: use the module gzip which is very similar to the standard file management method ([more info here](https://docs.python.org/3/library/gzip.html?highlight=gzip#module-gzip)). You can file a text file here [textFile.txt](file_samples/textFile.txt) and its gzipped version here [text.gz](file_samples/textFile.gz):\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import argparse\n",
    "import gzip\n",
    "\n",
    "parser = argparse.ArgumentParser(description=\"\"\"Reads and prints a text file\"\"\")\n",
    "parser.add_argument(\"filename\", type=str, help=\"The file name\")\n",
    "parser.add_argument(\"-z\", \"--gzipped\", action=\"store_true\", help=\"If set, input file is assumed gzipped\")\n",
    "\n",
    "args = parser.parse_args()\n",
    "inputFile = args.filename\n",
    "fh = \"\"\n",
    "if(args.gzipped):\n",
    "    fh = gzip.open(inputFile, \"rt\")\n",
    "else:\n",
    "    fh = open(inputFile, \"r\")\n",
    "\n",
    "for line in fh:\n",
    "    line = line.strip(\"\\n\")\n",
    "    print(line)\n",
    "\n",
    "fh.close()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The output:\n",
    "\n",
    "![](img/pract7/read_gz.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Exercises\n",
    "\n",
    "1. Modify the program of Exercise 4 of Practical 6 in order to allow users to specify the input and output files from command line. Then test it with the provided files. The text of the exercise follows:\n",
    "\n",
    "Write a python program that reads two files. The first is a one column text file ([contig_ids.txt](file_samples/contig_ids.txt)) with the identifiers of some contigs that are present in the second file, which is a fasta formatted file ([contigs82.fasta](file_samples/contigs82.fasta)). The program will write on a third, fasta formatted file (e.g. filtered_contigs.fasta) only those entries in *contigs82.fasta* having identifier in *contig_ids.txt*.\n",
    "\n",
    "\n",
    "\n",
    "<div class=\"tggle\" onclick=\"toggleVisibility('ex1');\">Show/Hide Solution</div>\n",
    "<div id=\"ex1\" style=\"display:none;\">"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "import argparse\n",
    "\n",
    "def readIDS(f):\n",
    "    \"\"\"reads a one column file in and stores\n",
    "    the ids in a dictionary that is returned at the end\"\"\"\n",
    "    ret = dict()\n",
    "    with open(f, \"r\") as file:\n",
    "        for line in file:\n",
    "            line = line.strip()\n",
    "            if(line not in ret):\n",
    "                ret[line] = 1\n",
    "    return ret\n",
    "\n",
    "def filterFasta(inF, outF, ids2keep):\n",
    "    oF = open(outF, \"w\")\n",
    "    \n",
    "    outputME = False\n",
    "    with open(inF, \"r\") as file:\n",
    "        for line in file:\n",
    "            line = line.strip()\n",
    "            if(line.startswith(\">\")):\n",
    "                #this is the header\n",
    "                if(line[1:] in ids2keep):\n",
    "                    oF.write(line +\"\\n\")\n",
    "                    outputME = True\n",
    "                    print(\"Writing contig \", line[1:])\n",
    "                else:\n",
    "                    outputME = False\n",
    "            else:\n",
    "                if(outputME):\n",
    "                    oF.write(line +\"\\n\")\n",
    "        \n",
    "    oF.close()\n",
    "    \n",
    "\n",
    "parser = argparse.ArgumentParser(description=\"Filters a fasta file\")\n",
    "parser.add_argument(\"inputFasta\", type = str, help = \"The input fasta file\")\n",
    "parser.add_argument(\"inputIDS\", type = str, help = \"The IDS to keep\")\n",
    "parser.add_argument(\"outputFasta\", type = str, help = \"The output fasta file with filtered entries\")\n",
    "args = parser.parse_args()\n",
    "idsFile = args.inputIDS\n",
    "inFasta = args.inputFasta\n",
    "outFasta = args.outputFasta\n",
    "\n",
    "ids = readIDS(idsFile)\n",
    "filterFasta(inFasta,outFasta, ids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "2. [Blast](https://www.ncbi.nlm.nih.gov/pubmed/2231712) is a well known tool to perform sequence alignment between a pool of query sequences and a pool of subject sequences. Among the other formats, it can produce an text output that is tab separated (```\\t```) capturing user specified output. Comments in the file are written in lines starting with an hash key (\"#\"). A sample blast output file is [blast_sample.tsv](file_samples/blast_sample.tsv), please download it and spend some time to look at it. The meaning of all columns is specified in the file header: \n",
    "```\n",
    "# Fields: query id, subject id, query length, % identity, alignment length, \n",
    "identical, gap opens, q. start, q. end, s. start, s. end, evalue\n",
    "```\n",
    "Write a python program with:\n",
    "    1. A function (*readBlast*) that reads in the blast .tsv file ignoring comment lines;\n",
    "    2. A function (*filterBlast*) that gets a string in input representing a blast alignment and filters it according to the following user specified parameters. It should return True if all filters are passed, false otherwise: \n",
    "    \n",
    "        1. % identity > identity_threshold (default 0%)\n",
    "        2. evalue < evalue_thrshold (default 0.5)\n",
    "        3. alignment length / query length > align_threshold (default 0)\n",
    "\n",
    "The program should report how many entries out of the total passed the filter.\n",
    "\n",
    "Test several combinations of filters like:\n",
    "\n",
    "    1. identity_theshold = 97, all others default\n",
    "    2. evalue_threshold = 0.5, all others default\n",
    "    3. align_threshold = 0.9, all others default\n",
    "    4. align_threshold = 0.9, identity_threshold = 97\n",
    "\n",
    "<div class=\"tggle\" onclick=\"toggleVisibility('ex3');\">Show/Hide Solution</div>\n",
    "<div id=\"ex3\" style=\"display:none;\">"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "8390 out of 90790 aligns passed filter (id_thr:97,eval_thr:1,al_thr:0)\n",
      "90780 out of 90790 aligns passed filter (id_thr:0,eval_thr:1,al_thr:0)\n",
      "27711 out of 90790 aligns passed filter (id_thr:0,eval_thr:1,al_thr:0.9)\n",
      "804 out of 90790 aligns passed filter (id_thr:97,eval_thr:1,al_thr:0.9)\n"
     ]
    }
   ],
   "source": [
    "\n",
    "\n",
    "def filterBlast(seq, id_thr = 0, eval_thr = 1, al_thr = 0):\n",
    "    \"\"\"filters an alignment represented as a string. Returns True if filters OK, False otherwise\"\"\"\n",
    "    infos = seq.split(\"\\t\")\n",
    "    return float(infos[3]) > id_thr and float(infos[-1]) < eval_thr and float(infos[4])/int(infos[2]) > al_thr\n",
    "\n",
    "def readBlast(inFile):\n",
    "    \"\"\"Reads the blast .tsv file and returns a list\"\"\"\n",
    "    fh = open(inFile, \"r\")\n",
    "    alignments = []\n",
    "    for line in fh:\n",
    "        if not line.startswith(\"#\"):\n",
    "            alignments.append(line)\n",
    "    return alignments\n",
    "\n",
    "def filterAndCount(alignList, id_thr = 0, eval_thr = 1, al_thr = 0):\n",
    "    \"\"\"filters alignments and counts the ones passing the filter\"\"\"\n",
    "    passed = [x for x in alignList if filterBlast(x, id_thr, eval_thr, al_thr)]\n",
    "    print(\"{} out of {} aligns passed filter (id_thr:{},eval_thr:{},al_thr:{})\".format(\n",
    "        len(passed), len(alignList), id_thr,eval_thr,al_thr))\n",
    "\n",
    "blastF = \"file_samples/blast_sample.tsv\"\n",
    "\n",
    "myAligns = readBlast(blastF)\n",
    "\n",
    "filterAndCount(myAligns, id_thr=97)\n",
    "filterAndCount(myAligns, eval_thr=1)\n",
    "filterAndCount(myAligns, al_thr=0.9)\n",
    "filterAndCount(myAligns, al_thr=0.9, id_thr=97)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "3. Fasta is a [format](https://en.wikipedia.org/wiki/FASTA_format) representing nucleotide or peptide sequences. \n",
    "Each entry of a fasta file starts with a \">\" followed by the identifier of the entry (the header of the sequence). All the lines following a header represent the sequence belonging to that identifier. Here you can find an example of fasta file: [contigs82.fasta](file_samples/contigs82.fasta). Please download it and have a look at its content.\n",
    "\n",
    "Write a python program that reads a fasta file and prints to screen the identifier of the sequence and the frequency of all the characters in the sequence (note that sequences might contain all IUPAC codes in case of SNPs). Hint: use a dictionary.\n",
    "\n",
    "<div class=\"tggle\" onclick=\"toggleVisibility('ex3');\">Show/Hide Solution</div>\n",
    "<div id=\"ex3\" style=\"display:none;\">"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MDC020656.85 :\n",
      "\t N has freq 0.052\n",
      "\t A has freq 0.308\n",
      "\t C has freq 0.163\n",
      "\t G has freq 0.203\n",
      "\t T has freq 0.273\n",
      "MDC001115.177 :\n",
      "\t N has freq 0.018\n",
      "\t T has freq 0.316\n",
      "\t C has freq 0.207\n",
      "\t G has freq 0.157\n",
      "\t A has freq 0.302\n",
      "MDC013284.379 :\n",
      "\t N has freq 0.005\n",
      "\t T has freq 0.321\n",
      "\t G has freq 0.194\n",
      "\t C has freq 0.171\n",
      "\t A has freq 0.309\n",
      "MDC018185.243 :\n",
      "\t N has freq 0.001\n",
      "\t W has freq 0.000\n",
      "\t R has freq 0.001\n",
      "\t Y has freq 0.001\n",
      "\t G has freq 0.181\n",
      "\t A has freq 0.307\n",
      "\t S has freq 0.000\n",
      "\t T has freq 0.304\n",
      "\t M has freq 0.000\n",
      "\t C has freq 0.204\n",
      "\t K has freq 0.000\n",
      "MDC018185.241 :\n",
      "\t N has freq 0.006\n",
      "\t W has freq 0.000\n",
      "\t R has freq 0.001\n",
      "\t Y has freq 0.002\n",
      "\t G has freq 0.183\n",
      "\t A has freq 0.335\n",
      "\t S has freq 0.000\n",
      "\t T has freq 0.304\n",
      "\t M has freq 0.001\n",
      "\t C has freq 0.168\n",
      "\t K has freq 0.001\n",
      "MDC004527.213 :\n",
      "\t N has freq 0.013\n",
      "\t T has freq 0.332\n",
      "\t G has freq 0.155\n",
      "\t C has freq 0.273\n",
      "\t A has freq 0.227\n",
      "MDC003661.174 :\n",
      "\t N has freq 0.007\n",
      "\t A has freq 0.246\n",
      "\t C has freq 0.213\n",
      "\t G has freq 0.232\n",
      "\t T has freq 0.301\n",
      "MDC012176.157 :\n",
      "\t N has freq 0.001\n",
      "\t A has freq 0.343\n",
      "\t C has freq 0.240\n",
      "\t G has freq 0.163\n",
      "\t T has freq 0.253\n",
      "MDC001204.812 :\n",
      "\t N has freq 0.000\n",
      "\t T has freq 0.298\n",
      "\t C has freq 0.221\n",
      "\t G has freq 0.187\n",
      "\t A has freq 0.293\n",
      "MDC001204.810 :\n",
      "\t N has freq 0.017\n",
      "\t A has freq 0.297\n",
      "\t C has freq 0.182\n",
      "\t G has freq 0.164\n",
      "\t T has freq 0.340\n",
      "MDC004389.256 :\n",
      "\t N has freq 0.028\n",
      "\t A has freq 0.285\n",
      "\t C has freq 0.277\n",
      "\t G has freq 0.207\n",
      "\t T has freq 0.203\n",
      "MDC024257.15 :\n",
      "\t N has freq 0.022\n",
      "\t T has freq 0.331\n",
      "\t G has freq 0.184\n",
      "\t C has freq 0.144\n",
      "\t A has freq 0.319\n",
      "MDC018297.229 :\n",
      "\t N has freq 0.031\n",
      "\t A has freq 0.275\n",
      "\t C has freq 0.181\n",
      "\t G has freq 0.209\n",
      "\t T has freq 0.305\n",
      "MDC001802.364 :\n",
      "\t N has freq 0.037\n",
      "\t A has freq 0.252\n",
      "\t C has freq 0.203\n",
      "\t G has freq 0.180\n",
      "\t T has freq 0.328\n",
      "MDC016621.241 :\n",
      "\t N has freq 0.063\n",
      "\t A has freq 0.302\n",
      "\t C has freq 0.206\n",
      "\t G has freq 0.168\n",
      "\t T has freq 0.262\n",
      "MDC014057.243 :\n",
      "\t N has freq 0.016\n",
      "\t A has freq 0.291\n",
      "\t G has freq 0.217\n",
      "\t C has freq 0.263\n",
      "\t T has freq 0.212\n",
      "MDC021015.302 :\n",
      "\t N has freq 0.081\n",
      "\t A has freq 0.302\n",
      "\t C has freq 0.143\n",
      "\t G has freq 0.251\n",
      "\t T has freq 0.223\n",
      "MDC018185.242 :\n",
      "\t N has freq 0.000\n",
      "\t S has freq 0.000\n",
      "\t R has freq 0.002\n",
      "\t Y has freq 0.001\n",
      "\t G has freq 0.189\n",
      "\t A has freq 0.316\n",
      "\t W has freq 0.000\n",
      "\t T has freq 0.289\n",
      "\t M has freq 0.001\n",
      "\t C has freq 0.200\n",
      "\t K has freq 0.000\n",
      "MDC051782.000 :\n",
      "\t N has freq 0.044\n",
      "\t T has freq 0.306\n",
      "\t S has freq 0.001\n",
      "\t W has freq 0.001\n",
      "\t G has freq 0.177\n",
      "\t Y has freq 0.002\n",
      "\t A has freq 0.285\n",
      "\t R has freq 0.002\n",
      "\t M has freq 0.000\n",
      "\t C has freq 0.180\n",
      "\t K has freq 0.001\n",
      "MDC017187.314 :\n",
      "\t N has freq 0.003\n",
      "\t A has freq 0.289\n",
      "\t G has freq 0.225\n",
      "\t C has freq 0.206\n",
      "\t T has freq 0.276\n",
      "MDC017187.311 :\n",
      "\t N has freq 0.068\n",
      "\t S has freq 0.000\n",
      "\t W has freq 0.001\n",
      "\t Y has freq 0.001\n",
      "\t G has freq 0.185\n",
      "\t A has freq 0.248\n",
      "\t R has freq 0.002\n",
      "\t T has freq 0.315\n",
      "\t M has freq 0.000\n",
      "\t C has freq 0.179\n",
      "\t K has freq 0.000\n",
      "MDC012865.410 :\n",
      "\t N has freq 0.003\n",
      "\t A has freq 0.289\n",
      "\t C has freq 0.162\n",
      "\t G has freq 0.218\n",
      "\t T has freq 0.328\n",
      "MDC000427.83 :\n",
      "\t N has freq 0.010\n",
      "\t A has freq 0.392\n",
      "\t C has freq 0.165\n",
      "\t G has freq 0.159\n",
      "\t T has freq 0.274\n",
      "MDC017187.319 :\n",
      "\t N has freq 0.006\n",
      "\t A has freq 0.282\n",
      "\t C has freq 0.181\n",
      "\t G has freq 0.200\n",
      "\t T has freq 0.331\n",
      "MDC017187.318 :\n",
      "\t N has freq 0.001\n",
      "\t T has freq 0.305\n",
      "\t G has freq 0.232\n",
      "\t A has freq 0.275\n",
      "\t C has freq 0.187\n",
      "\t K has freq 0.000\n",
      "MDC004081.319 :\n",
      "\t A has freq 0.172\n",
      "\t T has freq 0.197\n",
      "\t G has freq 0.098\n",
      "\t C has freq 0.106\n",
      "\t N has freq 0.428\n",
      "MDC021913.275 :\n",
      "\t N has freq 0.334\n",
      "\t A has freq 0.199\n",
      "\t C has freq 0.155\n",
      "\t G has freq 0.143\n",
      "\t T has freq 0.169\n",
      "MDC015147.205 :\n",
      "\t N has freq 0.002\n",
      "\t A has freq 0.298\n",
      "\t C has freq 0.154\n",
      "\t G has freq 0.196\n",
      "\t T has freq 0.350\n",
      "MDC000038.355 :\n",
      "\t N has freq 0.025\n",
      "\t W has freq 0.000\n",
      "\t R has freq 0.000\n",
      "\t Y has freq 0.000\n",
      "\t G has freq 0.158\n",
      "\t A has freq 0.331\n",
      "\t S has freq 0.000\n",
      "\t T has freq 0.315\n",
      "\t M has freq 0.000\n",
      "\t C has freq 0.170\n",
      "\t K has freq 0.000\n",
      "MDC016032.95 :\n",
      "\t N has freq 0.003\n",
      "\t R has freq 0.002\n",
      "\t A has freq 0.342\n",
      "\t G has freq 0.222\n",
      "\t T has freq 0.286\n",
      "\t Y has freq 0.001\n",
      "\t S has freq 0.001\n",
      "\t W has freq 0.001\n",
      "\t C has freq 0.142\n",
      "\t K has freq 0.002\n",
      "MDC052568.000 :\n",
      "\t N has freq 0.008\n",
      "\t A has freq 0.300\n",
      "\t S has freq 0.001\n",
      "\t W has freq 0.001\n",
      "\t G has freq 0.178\n",
      "\t Y has freq 0.001\n",
      "\t T has freq 0.302\n",
      "\t R has freq 0.002\n",
      "\t M has freq 0.001\n",
      "\t C has freq 0.207\n",
      "\t K has freq 0.000\n",
      "MDC008119.414 :\n",
      "\t N has freq 0.001\n",
      "\t T has freq 0.316\n",
      "\t C has freq 0.245\n",
      "\t G has freq 0.229\n",
      "\t A has freq 0.209\n",
      "MDC026201.7 :\n",
      "\t A has freq 0.313\n",
      "\t C has freq 0.247\n",
      "\t G has freq 0.191\n",
      "\t T has freq 0.250\n",
      "MDC003995.601 :\n",
      "\t N has freq 0.006\n",
      "\t W has freq 0.001\n",
      "\t A has freq 0.323\n",
      "\t G has freq 0.167\n",
      "\t T has freq 0.334\n",
      "\t Y has freq 0.001\n",
      "\t S has freq 0.000\n",
      "\t M has freq 0.001\n",
      "\t R has freq 0.001\n",
      "\t C has freq 0.166\n",
      "\t K has freq 0.000\n",
      "MDC009211.561 :\n",
      "\t N has freq 0.001\n",
      "\t T has freq 0.286\n",
      "\t G has freq 0.171\n",
      "\t C has freq 0.237\n",
      "\t A has freq 0.306\n",
      "MDC009211.567 :\n",
      "\t N has freq 0.006\n",
      "\t T has freq 0.327\n",
      "\t G has freq 0.181\n",
      "\t C has freq 0.167\n",
      "\t A has freq 0.318\n",
      "MDC054294.001 :\n",
      "\t A has freq 0.303\n",
      "\t S has freq 0.000\n",
      "\t W has freq 0.001\n",
      "\t Y has freq 0.002\n",
      "\t G has freq 0.186\n",
      "\t T has freq 0.318\n",
      "\t R has freq 0.002\n",
      "\t M has freq 0.001\n",
      "\t C has freq 0.186\n",
      "\t K has freq 0.001\n",
      "MDC004364.265 :\n",
      "\t N has freq 0.007\n",
      "\t W has freq 0.000\n",
      "\t R has freq 0.001\n",
      "\t Y has freq 0.001\n",
      "\t G has freq 0.198\n",
      "\t A has freq 0.323\n",
      "\t S has freq 0.000\n",
      "\t T has freq 0.302\n",
      "\t C has freq 0.170\n",
      "\t K has freq 0.000\n",
      "MDC002360.219 :\n",
      "\t N has freq 0.003\n",
      "\t A has freq 0.316\n",
      "\t C has freq 0.204\n",
      "\t G has freq 0.173\n",
      "\t T has freq 0.303\n",
      "MDC003408.117 :\n",
      "\t N has freq 0.017\n",
      "\t A has freq 0.329\n",
      "\t G has freq 0.191\n",
      "\t C has freq 0.215\n",
      "\t T has freq 0.247\n",
      "MDC015155.172 :\n",
      "\t N has freq 0.003\n",
      "\t T has freq 0.277\n",
      "\t G has freq 0.247\n",
      "\t C has freq 0.202\n",
      "\t A has freq 0.272\n",
      "MDC053310.000 :\n",
      "\t T has freq 0.310\n",
      "\t S has freq 0.000\n",
      "\t W has freq 0.001\n",
      "\t Y has freq 0.002\n",
      "\t G has freq 0.185\n",
      "\t A has freq 0.330\n",
      "\t R has freq 0.003\n",
      "\t M has freq 0.001\n",
      "\t C has freq 0.167\n",
      "\t K has freq 0.001\n",
      "MDC019140.398 :\n",
      "\t N has freq 0.005\n",
      "\t T has freq 0.302\n",
      "\t C has freq 0.163\n",
      "\t G has freq 0.178\n",
      "\t A has freq 0.352\n",
      "MDC019140.399 :\n",
      "\t N has freq 0.001\n",
      "\t A has freq 0.301\n",
      "\t C has freq 0.165\n",
      "\t G has freq 0.187\n",
      "\t T has freq 0.346\n",
      "MDC011390.337 :\n",
      "\t N has freq 0.046\n",
      "\t R has freq 0.000\n",
      "\t Y has freq 0.000\n",
      "\t G has freq 0.177\n",
      "\t A has freq 0.251\n",
      "\t T has freq 0.298\n",
      "\t C has freq 0.228\n",
      "MDC007154.375 :\n",
      "\t N has freq 0.003\n",
      "\t T has freq 0.292\n",
      "\t G has freq 0.201\n",
      "\t C has freq 0.197\n",
      "\t A has freq 0.307\n",
      "MDC006346.716 :\n",
      "\t N has freq 0.024\n",
      "\t A has freq 0.296\n",
      "\t C has freq 0.199\n",
      "\t G has freq 0.191\n",
      "\t T has freq 0.290\n",
      "MDC010588.505 :\n",
      "\t N has freq 0.006\n",
      "\t A has freq 0.305\n",
      "\t G has freq 0.195\n",
      "\t C has freq 0.257\n",
      "\t T has freq 0.237\n",
      "MDC002519.240 :\n",
      "\t N has freq 0.001\n",
      "\t T has freq 0.321\n",
      "\t R has freq 0.000\n",
      "\t Y has freq 0.000\n",
      "\t G has freq 0.188\n",
      "\t A has freq 0.312\n",
      "\t M has freq 0.000\n",
      "\t C has freq 0.178\n",
      "MDC031322.5 :\n",
      "\t N has freq 0.002\n",
      "\t W has freq 0.000\n",
      "\t R has freq 0.002\n",
      "\t A has freq 0.342\n",
      "\t G has freq 0.202\n",
      "\t T has freq 0.299\n",
      "\t Y has freq 0.002\n",
      "\t S has freq 0.000\n",
      "\t M has freq 0.001\n",
      "\t C has freq 0.149\n",
      "\t K has freq 0.000\n",
      "MDC010588.502 :\n",
      "\t N has freq 0.009\n",
      "\t T has freq 0.311\n",
      "\t C has freq 0.184\n",
      "\t G has freq 0.215\n",
      "\t A has freq 0.282\n",
      "MDC006346.711 :\n",
      "\t N has freq 0.006\n",
      "\t A has freq 0.321\n",
      "\t G has freq 0.187\n",
      "\t T has freq 0.295\n",
      "\t M has freq 0.000\n",
      "\t C has freq 0.191\n",
      "MDC011551.182 :\n",
      "\t N has freq 0.007\n",
      "\t T has freq 0.297\n",
      "\t W has freq 0.000\n",
      "\t A has freq 0.313\n",
      "\t G has freq 0.190\n",
      "\t S has freq 0.000\n",
      "\t Y has freq 0.001\n",
      "\t R has freq 0.001\n",
      "\t M has freq 0.000\n",
      "\t C has freq 0.189\n",
      "\t K has freq 0.000\n",
      "MDC002717.156 :\n",
      "\t N has freq 0.003\n",
      "\t T has freq 0.284\n",
      "\t C has freq 0.219\n",
      "\t G has freq 0.196\n",
      "\t A has freq 0.298\n",
      "MDC006346.719 :\n",
      "\t N has freq 0.006\n",
      "\t A has freq 0.286\n",
      "\t G has freq 0.219\n",
      "\t C has freq 0.183\n",
      "\t T has freq 0.306\n",
      "MDC007838.447 :\n",
      "\t N has freq 0.009\n",
      "\t T has freq 0.299\n",
      "\t C has freq 0.200\n",
      "\t G has freq 0.166\n",
      "\t A has freq 0.326\n",
      "MDC007018.186 :\n",
      "\t N has freq 0.151\n",
      "\t A has freq 0.226\n",
      "\t G has freq 0.202\n",
      "\t C has freq 0.160\n",
      "\t T has freq 0.262\n",
      "MDC017873.233 :\n",
      "\t N has freq 0.235\n",
      "\t A has freq 0.262\n",
      "\t C has freq 0.133\n",
      "\t G has freq 0.139\n",
      "\t T has freq 0.231\n",
      "MDC016296.138 :\n",
      "\t N has freq 0.016\n",
      "\t T has freq 0.346\n",
      "\t C has freq 0.172\n",
      "\t G has freq 0.188\n",
      "\t A has freq 0.277\n",
      "MDC019067.226 :\n",
      "\t N has freq 0.007\n",
      "\t A has freq 0.293\n",
      "\t C has freq 0.144\n",
      "\t G has freq 0.181\n",
      "\t T has freq 0.376\n",
      "MDC036568.1 :\n",
      "\t N has freq 0.006\n",
      "\t T has freq 0.348\n",
      "\t C has freq 0.154\n",
      "\t G has freq 0.161\n",
      "\t A has freq 0.331\n",
      "MDC014019.318 :\n",
      "\t N has freq 0.002\n",
      "\t A has freq 0.221\n",
      "\t G has freq 0.209\n",
      "\t C has freq 0.249\n",
      "\t T has freq 0.319\n",
      "MDC007995.528 :\n",
      "\t N has freq 0.009\n",
      "\t T has freq 0.340\n",
      "\t C has freq 0.166\n",
      "\t G has freq 0.181\n",
      "\t A has freq 0.305\n",
      "MDC026961.60 :\n",
      "\t N has freq 0.005\n",
      "\t A has freq 0.300\n",
      "\t C has freq 0.201\n",
      "\t G has freq 0.206\n",
      "\t T has freq 0.288\n",
      "MDC013443.168 :\n",
      "\t N has freq 0.005\n",
      "\t A has freq 0.289\n",
      "\t G has freq 0.173\n",
      "\t C has freq 0.203\n",
      "\t T has freq 0.330\n",
      "MDC022800.298 :\n",
      "\t N has freq 0.008\n",
      "\t A has freq 0.323\n",
      "\t G has freq 0.203\n",
      "\t C has freq 0.131\n",
      "\t T has freq 0.334\n",
      "MDC021558.159 :\n",
      "\t N has freq 0.002\n",
      "\t A has freq 0.298\n",
      "\t C has freq 0.176\n",
      "\t G has freq 0.190\n",
      "\t T has freq 0.334\n",
      "MDC002479.192 :\n",
      "\t N has freq 0.009\n",
      "\t T has freq 0.304\n",
      "\t G has freq 0.232\n",
      "\t C has freq 0.151\n",
      "\t A has freq 0.305\n",
      "MDC010751.301 :\n",
      "\t N has freq 0.006\n",
      "\t A has freq 0.318\n",
      "\t C has freq 0.172\n",
      "\t G has freq 0.214\n",
      "\t T has freq 0.291\n",
      "MDC000219.190 :\n",
      "\t N has freq 0.003\n",
      "\t R has freq 0.000\n",
      "\t Y has freq 0.001\n",
      "\t G has freq 0.168\n",
      "\t A has freq 0.332\n",
      "\t S has freq 0.000\n",
      "\t T has freq 0.335\n",
      "\t M has freq 0.000\n",
      "\t C has freq 0.160\n",
      "MDC020963.161 :\n",
      "\t N has freq 0.002\n",
      "\t A has freq 0.287\n",
      "\t C has freq 0.248\n",
      "\t G has freq 0.148\n",
      "\t T has freq 0.315\n",
      "MDC020963.162 :\n",
      "\t N has freq 0.005\n",
      "\t W has freq 0.000\n",
      "\t G has freq 0.207\n",
      "\t A has freq 0.309\n",
      "\t S has freq 0.000\n",
      "\t T has freq 0.265\n",
      "\t C has freq 0.214\n",
      "\t R has freq 0.000\n",
      "MDC051637.000 :\n",
      "\t N has freq 0.010\n",
      "\t T has freq 0.328\n",
      "\t W has freq 0.001\n",
      "\t R has freq 0.003\n",
      "\t Y has freq 0.005\n",
      "\t G has freq 0.169\n",
      "\t A has freq 0.307\n",
      "\t S has freq 0.000\n",
      "\t M has freq 0.001\n",
      "\t C has freq 0.175\n",
      "\t K has freq 0.001\n",
      "MDC005174.220 :\n",
      "\t N has freq 0.022\n",
      "\t A has freq 0.311\n",
      "\t G has freq 0.186\n",
      "\t C has freq 0.198\n",
      "\t T has freq 0.283\n",
      "MDC040033.7 :\n",
      "\t N has freq 0.003\n",
      "\t A has freq 0.297\n",
      "\t G has freq 0.204\n",
      "\t C has freq 0.195\n",
      "\t T has freq 0.301\n",
      "MDC019674.147 :\n",
      "\t N has freq 0.004\n",
      "\t W has freq 0.000\n",
      "\t A has freq 0.324\n",
      "\t G has freq 0.167\n",
      "\t T has freq 0.366\n",
      "\t Y has freq 0.001\n",
      "\t S has freq 0.000\n",
      "\t C has freq 0.137\n",
      "\t R has freq 0.000\n",
      "MDC010450.877 :\n",
      "\t N has freq 0.058\n",
      "\t A has freq 0.293\n",
      "\t G has freq 0.174\n",
      "\t C has freq 0.190\n",
      "\t T has freq 0.285\n",
      "MDC007097.457 :\n",
      "\t N has freq 0.070\n",
      "\t T has freq 0.299\n",
      "\t C has freq 0.126\n",
      "\t G has freq 0.217\n",
      "\t A has freq 0.289\n",
      "MDC016278.70 :\n",
      "\t N has freq 0.084\n",
      "\t A has freq 0.277\n",
      "\t C has freq 0.183\n",
      "\t G has freq 0.153\n",
      "\t T has freq 0.303\n",
      "MDC013723.254 :\n",
      "\t N has freq 0.001\n",
      "\t A has freq 0.279\n",
      "\t G has freq 0.179\n",
      "\t C has freq 0.212\n",
      "\t T has freq 0.329\n",
      "MDC002838.179 :\n",
      "\t N has freq 0.013\n",
      "\t A has freq 0.359\n",
      "\t G has freq 0.167\n",
      "\t C has freq 0.163\n",
      "\t T has freq 0.297\n",
      "MDC009771.217 :\n",
      "\t N has freq 0.001\n",
      "\t T has freq 0.335\n",
      "\t G has freq 0.163\n",
      "\t C has freq 0.185\n",
      "\t A has freq 0.315\n"
     ]
    }
   ],
   "source": [
    "def countFrequency(seq):\n",
    "    \"\"\"gets a sequence in input and returns\n",
    "    a dictionary with bases as keys and frequence as value\"\"\"\n",
    "    bases = {}\n",
    "    for b in seq:\n",
    "        if b not in bases:\n",
    "            bases[b] = 1\n",
    "        else:\n",
    "            bases[b] += 1\n",
    "    \n",
    "    for b in bases:\n",
    "        bases[b] = bases[b]/len(seq)\n",
    "    return bases\n",
    "\n",
    "def printData(ident, freqDict):\n",
    "    \"\"\"get the identifier and a dictionary with all frequencies\n",
    "    and prints both information on the screen\"\"\"\n",
    "    print(ident,\":\")\n",
    "    for f in freqDict:\n",
    "        print(\"\\t {} has freq {:.3f}\".format(f,freqDict[f]))\n",
    "\n",
    "def processFasta(file):\n",
    "    header = \"\"\n",
    "    seq = \"\"\n",
    "    with open(file, \"r\") as f:\n",
    "        for line in f:\n",
    "            line = line.strip()\n",
    "            if(line.startswith(\">\")):\n",
    "                if(len(header) == 0 ):\n",
    "                    #first entry:\n",
    "                    header = line[1:]\n",
    "                else:\n",
    "                    #this is a new entry\n",
    "                    freqData = countFrequency(seq)\n",
    "                    printData(header,freqData)\n",
    "                    seq = \"\"\n",
    "                    header = line[1:]\n",
    "            else:\n",
    "                seq +=line\n",
    "    #processing the final entry\n",
    "    freqData = countFrequency(seq)\n",
    "    printData(header,freqData)\n",
    "                    \n",
    "\n",
    "inFasta = \"file_samples/contigs82.fasta\"\n",
    "processFasta(inFasta)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "4. Write a python program that reads two files. The first is a one column text file ([contig_ids.txt](file_samples/contig_ids.txt)) with the identifiers of some contigs that are present in the second file, which is a fasta formatted file ([contigs82.fasta](file_samples/contigs82.fasta)). The program will write on a third, fasta formatted file (e.g. filtered_contigs.fasta) only those entries in *contigs82.fasta* having identifier in *contig_ids.txt*.\n",
    "\n",
    "<div class=\"tggle\" onclick=\"toggleVisibility('ex4');\">Show/Hide Solution</div>\n",
    "<div id=\"ex4\" style=\"display:none;\">"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing contig  MDC020656.85\n",
      "Writing contig  MDC001115.177\n",
      "Writing contig  MDC013284.379\n",
      "Writing contig  MDC018185.243\n",
      "Writing contig  MDC018185.241\n",
      "Writing contig  MDC004527.213\n",
      "Writing contig  MDC012176.157\n",
      "Writing contig  MDC001204.810\n",
      "Writing contig  MDC004389.256\n",
      "Writing contig  MDC018297.229\n",
      "Writing contig  MDC001802.364\n",
      "Writing contig  MDC014057.243\n",
      "Writing contig  MDC021015.302\n",
      "Writing contig  MDC017187.314\n",
      "Writing contig  MDC012865.410\n",
      "Writing contig  MDC000427.83\n",
      "Writing contig  MDC017187.319\n",
      "Writing contig  MDC004364.265\n",
      "Writing contig  MDC002360.219\n",
      "Writing contig  MDC015155.172\n",
      "Writing contig  MDC019140.398\n",
      "Writing contig  MDC019140.399\n",
      "Writing contig  MDC011390.337\n",
      "Writing contig  MDC007154.375\n",
      "Writing contig  MDC010588.505\n",
      "Writing contig  MDC002519.240\n",
      "Writing contig  MDC006346.711\n",
      "Writing contig  MDC011551.182\n",
      "Writing contig  MDC002717.156\n",
      "Writing contig  MDC006346.719\n",
      "Writing contig  MDC007838.447\n",
      "Writing contig  MDC007018.186\n",
      "Writing contig  MDC017873.233\n",
      "Writing contig  MDC016296.138\n",
      "Writing contig  MDC019067.226\n",
      "Writing contig  MDC014019.318\n",
      "Writing contig  MDC007995.528\n",
      "Writing contig  MDC026961.60\n",
      "Writing contig  MDC020963.161\n",
      "Writing contig  MDC005174.220\n",
      "Writing contig  MDC040033.7\n",
      "Writing contig  MDC019674.147\n",
      "Writing contig  MDC010450.877\n",
      "Writing contig  MDC007097.457\n",
      "Writing contig  MDC016278.70\n"
     ]
    }
   ],
   "source": [
    "def readIDS(f):\n",
    "    \"\"\"reads a one column file in and stores\n",
    "    the ids in a dictionary that is returned at the end\"\"\"\n",
    "    ret = dict()\n",
    "    with open(f, \"r\") as file:\n",
    "        for line in file:\n",
    "            line = line.strip()\n",
    "            if(line not in ret):\n",
    "                ret[line] = 1\n",
    "    return ret\n",
    "\n",
    "def filterFasta(inF, outF, ids2keep):\n",
    "    oF = open(outF, \"w\")\n",
    "    \n",
    "    outputME = False\n",
    "    with open(inF, \"r\") as file:\n",
    "        for line in file:\n",
    "            line = line.strip()\n",
    "            if(line.startswith(\">\")):\n",
    "                #this is the header\n",
    "                if(line[1:] in ids2keep):\n",
    "                    oF.write(line +\"\\n\")\n",
    "                    outputME = True\n",
    "                    print(\"Writing contig \", line[1:])\n",
    "                else:\n",
    "                    outputME = False\n",
    "            else:\n",
    "                if(outputME):\n",
    "                    oF.write(line +\"\\n\")\n",
    "        \n",
    "    oF.close()\n",
    "    \n",
    "\n",
    "idsFile = \"file_samples/contig_ids.txt\"\n",
    "inFasta = \"file_samples/contigs82.fasta\"\n",
    "outFasta = \"file_samples/filtered_contigs.fasta\"\n",
    "\n",
    "ids = readIDS(idsFile)\n",
    "filterFasta(inFasta,outFasta, ids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "5. Write a python program that:\n",
    "\n",
    "    1. reads the text file [sample_text.txt](file_samples/sample_text.txt) and stores in a dictionary how many times each word appears (hint: the key is the word and the count is the value);\n",
    "    2. prints to screen how many lines the file has and how many distinct words are in the file;\n",
    "    3. writes to a text file (scientist_histo.csv) the histogram of the words in comma separated value format (i.e. word,count). Words must be sorted alphabetically;\n",
    "    4. Finally, write a function that prints to screen (alphabetically) all the words that have a count higher than a threshold N and apply it with N = 15.\n",
    "    \n",
    "<div class=\"tggle\" onclick=\"toggleVisibility('ex5');\">Show/Hide Solution</div>\n",
    "<div id=\"ex5\" style=\"display:none;\">"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The file file_samples/sample_text.txt has 28 lines\n",
      "The total number of distinct words is  308\n",
      "Word \"challenge\" is present 6 times\n",
      "Word \"future\" is present 22 times\n",
      "Word \"reply\" is present 7 times\n",
      "Word \"shop\" is present 6 times\n",
      "Word \"thunder\" is present 11 times\n",
      "Word \"umbrella\" is present 10 times\n"
     ]
    }
   ],
   "source": [
    "def wordHisto(myText):\n",
    "    \"\"\"this function returns a dictionary\n",
    "    with the count of each word in myText (separated by \" \" or \"\\n\") \n",
    "    \"\"\"\n",
    "    myDict = dict()\n",
    "    tmp = myText.replace(\"\\n\",\" \")\n",
    "    \n",
    "    for word in tmp.split(\" \"):\n",
    "        if(word not in myDict):\n",
    "            myDict[word] = 1\n",
    "        else:\n",
    "            myDict[word] += 1\n",
    "    return myDict\n",
    "\n",
    "def writeWordHisto(outF, data):\n",
    "    \"\"\"this function writes to outFile\n",
    "    the word histogram data contained in the dictionary data\n",
    "    the output format is comma separated\n",
    "    \"\"\"\n",
    "    fh = open(outF, \"w\")\n",
    "    dictKeys = list(data.keys())\n",
    "    dictKeys.sort()\n",
    "    fh.write(\"#word,count\\n\")\n",
    "    for k in dictKeys:\n",
    "        curVal = data[k]\n",
    "        myStr = k + \",\" + str(curVal)+\"\\n\" #string to write in file\n",
    "        fh.write(myStr)\n",
    "    fh.close() #remember to close the file\n",
    "\n",
    "def printWords(N, data):\n",
    "    \"\"\"prints the words in data that have a count higher than N\"\"\"\n",
    "    dictKeys = list(data.keys())\n",
    "    dictKeys.sort()\n",
    "    for w in dictKeys:\n",
    "        cnt = data[w]\n",
    "        if(cnt > N):\n",
    "            print(\"Word \\\"{}\\\" is present {} times\".format( w, cnt))\n",
    "    \n",
    "    \n",
    "file = \"file_samples/sample_text.txt\"\n",
    "outFile = \"file_samples/sample_text_histo.csv\"\n",
    "wholeText = \"\"\n",
    "fh = open(file,\"r\")\n",
    "wholeText = fh.read()\n",
    "wholeText = wholeText.strip() #to remove the final newline character\n",
    "\n",
    "fh.close()\n",
    "print(\"The file {} has {} lines\".format(file, wholeText.count(\"\\n\") +1 )) #n lines have n-1 \\n\n",
    "\n",
    "#Let's do the job:\n",
    "wordD = wordHisto(wholeText)\n",
    "writeWordHisto(outFile, wordD)\n",
    "print(\"The total number of distinct words is \", len(wordD))\n",
    "printWords(5,wordD)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "6. Write the following python functions and test them with some parameters of your choice: \n",
    "\n",
    "    1. *getDivisors*: the function has a positive integer as parameter and returns a list of all the positive divisors of the integer in input (excluding the number itself). Example: ```getDivisors(6) --> [1,2,3]```\n",
    "    \n",
    "    2. *checkSum*: the function has a list and an integer as parameters and returns True if the sum of all elements in the list  equals the integer, False otherwise. Example: ```checkSum([1,2,3], 6) --> True```, ```checkSum([1,2,3],1) --> False```.\n",
    "    \n",
    "    3. *checkPerfect*: the function gets an integer as parameter and returns True if the integer is a [perfect number](https://en.wikipedia.org/wiki/Perfect_number), False otherwise. A number is perfect if all its divisors (excluding itself) sum to its value. Example: ```checkPerfect(6) --> True``` because 1+2+3 = 6. Hint: use the functions implemented before.\n",
    "    \n",
    "Use the three implemented functions to write a fourth function:\n",
    "\n",
    "*getFirstNperfects*: the function gets an integer N as parameter and returns a dictionary with the first N perfect numbers. The key of the dictionary is the perfect number, while the value of the dictionary is the list of its divisors. Example: ```getFirstNperfects(1) --> {6 : [1,2,3]}```\n",
    "    \n",
    "Get and print the first 4 perfect numbers and finally test if 33550336 is a perfect number.\n",
    "\n",
    "**WARNING:** do not try to find more than 4 perfect numbers as it might take a while!!!\n",
    "\n",
    "<div class=\"tggle\" onclick=\"toggleVisibility('ex6');\">Show/Hide Solution</div>\n",
    "<div id=\"ex6\" style=\"display:none;\">"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "6  =  1+2+3\n",
      "28  =  1+2+4+7+14\n",
      "496  =  1+2+4+8+16+31+62+124+248\n",
      "8128  =  1+2+4+8+16+32+64+127+254+508+1016+2032+4064\n",
      "Is 33550336 a perfect number? True\n"
     ]
    }
   ],
   "source": [
    "def getDivisors(intVal):\n",
    "    \"\"\"returns the integer divisors of intVal\"\"\"\n",
    "    ret = [x for x in range(1,intVal//2 + 1) if intVal % x == 0]\n",
    "    #OR:\n",
    "    #for i in range(1,intVal//2+1):\n",
    "    #    if(intVal % i == 0):\n",
    "    #        ret.append(i)\n",
    "    return ret\n",
    "\n",
    "def checkSum(intList, intVal):\n",
    "    \"\"\"checks if the sum of elements in intList equals intVal\"\"\"\n",
    "    s = 0\n",
    "    for x in intList:\n",
    "        s += x\n",
    "    return (s == intVal)\n",
    "\n",
    "def checkPerfect(intVal):\n",
    "    \"\"\"checks if intVal is a perfect number\"\"\"\n",
    "    divisors = getDivisors(intVal)\n",
    "    return checkSum(divisors,intVal)\n",
    "\n",
    "def getFirstNPerfects(N):\n",
    "    \"\"\"Finds the first N perfect numbers\"\"\"\n",
    "    i = 0\n",
    "    val = 2\n",
    "    ret = {}\n",
    "    while(i<N):\n",
    "        if(checkPerfect(val)):\n",
    "            i+=1\n",
    "            ret[val] = getDivisors(val)\n",
    "            val += 1\n",
    "        else:\n",
    "            val += 1\n",
    "    \n",
    "    return ret\n",
    "            \n",
    "        \n",
    "perfects = getFirstNPerfects(4)\n",
    "perKeys = list(perfects.keys())\n",
    "perKeys.sort()\n",
    "for p in perKeys:\n",
    "    print(p, \" = \", \"+\".join([str(x) for x in perfects[p]]))\n",
    "    \n",
    "print(\"Is 33550336 a perfect number?\", checkPerfect(33550336))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "</div>"
   ]
  }
 ],
 "metadata": {
  "celltoolbar": "Edit Metadata",
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
